{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# COMP90014 Assignment 2\n",
    "### Semester 2, 2020\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This assignment should be completed by each student individually. Make sure you read this entire document, and ask for help if anything is not clear. Any changes or clarification to this document will be announced via Canvas.\n",
    "\n",
    "Please make sure you are aware of the University's rules on academic honesty and plagiarism, which are very strict: https://academichonesty.unimelb.edu.au/ \n",
    "\n",
    "Make sure you **do not** copy any code either from other students or from the internet. This is considered plagiarism. It is generally a good idea to avoid looking at any solutions as you may find it surprisingly difficult to generate your own solution to the problem once you have seen somebody else's.\n",
    "\n",
    "Your completed notebook file containing all your answers will be turned in via Canvas. No other files or formats will be accepted - only upload the completed `.ipynb` file."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Overview\n",
    "To complete the assignment you will need to finish the tasks in this notebook. There are multiple tasks that are connected in a logical order.\n",
    "\n",
    "The tasks are a combination of writing your own implementations of algorithms and data structures we've discussed in lectures (including adapting or describing them in plain English or pseudocode), interpret and analyse results  and your implementation choices. Each answer question has a word limit that will be strictly enforced! **Please note that for this assignment, you may be awarded a mark of zero for a question if you go over the word limit.**\n",
    "\n",
    "In some case, we have provided test input and test output that you can use to try out your solutions. These tests are just samples and are not exhaustive - they may warn you if you've made a mistake, but they are not guaranteed to. It's up to you to decide whether your code is correct."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pseudocode\n",
    "Pseudocode for algorithms are a series of logical steps which any programmer can understand.\n",
    "Here is the pseudocode for a function fizzbuzz to print numbers that are divisible by 3 or 5:\n",
    "\n",
    "```\n",
    "function fizzbuzz()\n",
    "    For i = 1 to 100\n",
    "        If i is divisible by 3 Then\n",
    "            Print \"Fizz\"\n",
    "        If i is divisible by 5 Then\n",
    "            Print \"Buzz\"\n",
    "```\n",
    "      \n",
    "As you can see, the basic steps are shown, but there is no language-specific syntax. <br>\n",
    "In this manner, pseudocode explains the algorithm procedure in **direct, plain language.** <br>\n",
    "As a note, if your function calls another function (lets call this function **boom**), write it as **boom()** in the pseudocode. The open/close brackets show that 'boom' is another function call.\n",
    "\n",
    "There are no real conventions aside from the above, so please use a style which you think is **clearest.**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Written Answer Questions\n",
    "\n",
    "A portion of this assignment is written answer. <br>\n",
    "We ask for students to use bullet points when possible. Bullet points clearly outline the order of steps which need to be performed. <br>\n",
    "\n",
    "Here is an example:<br>\n",
    "**Question:** <br>\n",
    "Describe with your own words how you would implement a function to calculate the GC content of a DNA sequence <br>\n",
    "**Answer:** <br>\n",
    "The GC content is the proportion of nucleotides that are either guanine ('G') or cytosine('C') in a sequence. To calculate:\n",
    "* record the length of the sequence\n",
    "* set a variable (call it GC_count) to 0 which will store the number of G and C bases\n",
    "* for each nucleotide in the sequence:\n",
    "    * if the nucleotide is 'G' or 'C', increment GC_count by 1\n",
    "* return GC_count / sequence length * 100\n",
    "* This will return the GC content as a percentage of all nucleotides\n",
    "\n",
    "These written answers don't need to be as detailed as pseudocode, where you should note any variables you declare, functions you call etc. <br>\n",
    "Written answers should be interpretable by a bioinformatician, who could then understand your method and write code if necessary. <br>\n",
    "The broad ideas and logical steps need to be there, but the exactly implementation does not always have to be written. <br>\n",
    "Pseudocode is a little different, and should be addressed as per the cell above."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Marking\n",
    "\n",
    "Cells that must be completed to receive marks are clearly labeled. There are 14 graded cells, some of which are code cells, in which you must complete the code to solve a problem, and some of which are markdown cells, in which you must write your answers to questions or descriptions of algorithmic approaches (pseudocode or plain English might be requested). In order to get full marks, your code/answer needs to be clear, unambiguous, with adequate level of detail (and comments for Python code) as well as optimal time complexity. **Code style, readability and efficiency** will be taken into consideration.\n",
    "\n",
    "The total marks for the assignment add up to 40, and it will be worth 15% of your overall subject grade."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part I\n",
    "\n",
    "### Background and data \n",
    "\n",
    "WGCNA stands for weighted gene co-expression network analysis. It is a data analysis technique used for studying biological networks based on pairwise correlations of gene expression data. WGCNA is good at identifying clusters of genes that may be co-regulated, and therefore may have shared biological function.\n",
    "\n",
    "For this assignment, you will primarily be using the [FlyAtlas](http://flyatlas.org) dataset. For this assignment, instead of using the probe-wise dataset, we will be using the expression value for each gene.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Setup "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import altair"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import networkx as nx\n",
    "import scipy\n",
    "import re\n",
    "from io import StringIO\n",
    "from sklearn.decomposition import PCA\n",
    "from scipy.cluster.hierarchy import linkage, dendrogram, fcluster\n",
    "from scipy.spatial.distance import squareform"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read in data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "raw_expression = pd.read_csv('flyatlas_subset.csv.gz', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Adult Hind Gut biological rep1</th>\n",
       "      <th>Adult Hind Gut biological rep2</th>\n",
       "      <th>Adult Hind Gut biological rep3</th>\n",
       "      <th>Adult Hind Gut biological rep4</th>\n",
       "      <th>Adult Mid Gut biological rep1</th>\n",
       "      <th>Adult Mid Gut biological rep2</th>\n",
       "      <th>Adult Mid Gut biological rep3</th>\n",
       "      <th>Adult Mid Gut biological rep4</th>\n",
       "      <th>Adult Accessory gland biological rep1</th>\n",
       "      <th>Adult Accessory gland biological rep2</th>\n",
       "      <th>...</th>\n",
       "      <th>Adult Thoracic Muscle biological rep3</th>\n",
       "      <th>Adult Thoracic Muscle biological rep4</th>\n",
       "      <th>Adult Wings biological rep1</th>\n",
       "      <th>Adult Wings biological rep2</th>\n",
       "      <th>Adult Wings biological rep3</th>\n",
       "      <th>Adult Wings biological rep4</th>\n",
       "      <th>Adult Female Spermatheca Mated (REDONE) biological rep1</th>\n",
       "      <th>Adult Female Spermatheca Mated (REDONE) biological rep2</th>\n",
       "      <th>Adult Female Spermatheca Mated (REDONE) biological rep3</th>\n",
       "      <th>Adult Female Spermatheca Mated (REDONE) biological rep4\"\\n</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>128up</th>\n",
       "      <td>214.51500</td>\n",
       "      <td>246.59700</td>\n",
       "      <td>216.58100</td>\n",
       "      <td>221.66700</td>\n",
       "      <td>264.602000</td>\n",
       "      <td>281.02000</td>\n",
       "      <td>300.132000</td>\n",
       "      <td>255.19500</td>\n",
       "      <td>427.68900</td>\n",
       "      <td>467.309000</td>\n",
       "      <td>...</td>\n",
       "      <td>113.9360</td>\n",
       "      <td>363.59600</td>\n",
       "      <td>196.98400</td>\n",
       "      <td>262.50700</td>\n",
       "      <td>181.05800</td>\n",
       "      <td>204.36200</td>\n",
       "      <td>380.878000</td>\n",
       "      <td>112.6860</td>\n",
       "      <td>294.49000</td>\n",
       "      <td>329.094000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140up</th>\n",
       "      <td>139.86300</td>\n",
       "      <td>181.65600</td>\n",
       "      <td>143.80200</td>\n",
       "      <td>149.44800</td>\n",
       "      <td>92.043300</td>\n",
       "      <td>82.10540</td>\n",
       "      <td>77.407200</td>\n",
       "      <td>91.44190</td>\n",
       "      <td>116.91800</td>\n",
       "      <td>124.892000</td>\n",
       "      <td>...</td>\n",
       "      <td>155.8030</td>\n",
       "      <td>110.83100</td>\n",
       "      <td>79.40570</td>\n",
       "      <td>81.80540</td>\n",
       "      <td>76.40320</td>\n",
       "      <td>68.40710</td>\n",
       "      <td>85.908800</td>\n",
       "      <td>128.3340</td>\n",
       "      <td>87.18940</td>\n",
       "      <td>104.890000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18w</th>\n",
       "      <td>241.72200</td>\n",
       "      <td>247.02900</td>\n",
       "      <td>245.75500</td>\n",
       "      <td>262.18900</td>\n",
       "      <td>54.108400</td>\n",
       "      <td>60.05170</td>\n",
       "      <td>60.698900</td>\n",
       "      <td>54.93600</td>\n",
       "      <td>33.02630</td>\n",
       "      <td>39.497000</td>\n",
       "      <td>...</td>\n",
       "      <td>100.4050</td>\n",
       "      <td>294.09100</td>\n",
       "      <td>772.38800</td>\n",
       "      <td>700.44800</td>\n",
       "      <td>752.14300</td>\n",
       "      <td>754.71500</td>\n",
       "      <td>285.764000</td>\n",
       "      <td>66.5310</td>\n",
       "      <td>248.04200</td>\n",
       "      <td>264.250000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>312</th>\n",
       "      <td>4.16597</td>\n",
       "      <td>2.60007</td>\n",
       "      <td>3.53377</td>\n",
       "      <td>4.60587</td>\n",
       "      <td>0.194508</td>\n",
       "      <td>0.57742</td>\n",
       "      <td>0.977839</td>\n",
       "      <td>2.65123</td>\n",
       "      <td>8.84807</td>\n",
       "      <td>0.342770</td>\n",
       "      <td>...</td>\n",
       "      <td>68.3797</td>\n",
       "      <td>0.36183</td>\n",
       "      <td>7.03675</td>\n",
       "      <td>2.94253</td>\n",
       "      <td>8.25821</td>\n",
       "      <td>5.44387</td>\n",
       "      <td>5.008850</td>\n",
       "      <td>35.0969</td>\n",
       "      <td>2.39418</td>\n",
       "      <td>2.270310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7B2</th>\n",
       "      <td>25.78430</td>\n",
       "      <td>19.12000</td>\n",
       "      <td>34.80000</td>\n",
       "      <td>21.64910</td>\n",
       "      <td>89.001700</td>\n",
       "      <td>116.57100</td>\n",
       "      <td>114.170000</td>\n",
       "      <td>123.40500</td>\n",
       "      <td>2.34292</td>\n",
       "      <td>0.618435</td>\n",
       "      <td>...</td>\n",
       "      <td>175.8510</td>\n",
       "      <td>1.38861</td>\n",
       "      <td>172.11500</td>\n",
       "      <td>194.96700</td>\n",
       "      <td>128.78000</td>\n",
       "      <td>183.37900</td>\n",
       "      <td>0.842605</td>\n",
       "      <td>16.0162</td>\n",
       "      <td>2.64286</td>\n",
       "      <td>0.579274</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 136 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Adult Hind Gut biological rep1  Adult Hind Gut biological rep2  \\\n",
       "Gene Symbol                                                                   \n",
       "128up                             214.51500                       246.59700   \n",
       "140up                             139.86300                       181.65600   \n",
       "18w                               241.72200                       247.02900   \n",
       "312                                 4.16597                         2.60007   \n",
       "7B2                                25.78430                        19.12000   \n",
       "\n",
       "             Adult Hind Gut biological rep3  Adult Hind Gut biological rep4  \\\n",
       "Gene Symbol                                                                   \n",
       "128up                             216.58100                       221.66700   \n",
       "140up                             143.80200                       149.44800   \n",
       "18w                               245.75500                       262.18900   \n",
       "312                                 3.53377                         4.60587   \n",
       "7B2                                34.80000                        21.64910   \n",
       "\n",
       "             Adult Mid Gut biological rep1  Adult Mid Gut biological rep2  \\\n",
       "Gene Symbol                                                                 \n",
       "128up                           264.602000                      281.02000   \n",
       "140up                            92.043300                       82.10540   \n",
       "18w                              54.108400                       60.05170   \n",
       "312                               0.194508                        0.57742   \n",
       "7B2                              89.001700                      116.57100   \n",
       "\n",
       "             Adult Mid Gut biological rep3  Adult Mid Gut biological rep4  \\\n",
       "Gene Symbol                                                                 \n",
       "128up                           300.132000                      255.19500   \n",
       "140up                            77.407200                       91.44190   \n",
       "18w                              60.698900                       54.93600   \n",
       "312                               0.977839                        2.65123   \n",
       "7B2                             114.170000                      123.40500   \n",
       "\n",
       "             Adult Accessory gland biological rep1  \\\n",
       "Gene Symbol                                          \n",
       "128up                                    427.68900   \n",
       "140up                                    116.91800   \n",
       "18w                                       33.02630   \n",
       "312                                        8.84807   \n",
       "7B2                                        2.34292   \n",
       "\n",
       "             Adult Accessory gland biological rep2  ...  \\\n",
       "Gene Symbol                                         ...   \n",
       "128up                                   467.309000  ...   \n",
       "140up                                   124.892000  ...   \n",
       "18w                                      39.497000  ...   \n",
       "312                                       0.342770  ...   \n",
       "7B2                                       0.618435  ...   \n",
       "\n",
       "             Adult Thoracic Muscle biological rep3  \\\n",
       "Gene Symbol                                          \n",
       "128up                                     113.9360   \n",
       "140up                                     155.8030   \n",
       "18w                                       100.4050   \n",
       "312                                        68.3797   \n",
       "7B2                                       175.8510   \n",
       "\n",
       "             Adult Thoracic Muscle biological rep4  \\\n",
       "Gene Symbol                                          \n",
       "128up                                    363.59600   \n",
       "140up                                    110.83100   \n",
       "18w                                      294.09100   \n",
       "312                                        0.36183   \n",
       "7B2                                        1.38861   \n",
       "\n",
       "             Adult Wings biological rep1  Adult Wings biological rep2  \\\n",
       "Gene Symbol                                                             \n",
       "128up                          196.98400                    262.50700   \n",
       "140up                           79.40570                     81.80540   \n",
       "18w                            772.38800                    700.44800   \n",
       "312                              7.03675                      2.94253   \n",
       "7B2                            172.11500                    194.96700   \n",
       "\n",
       "             Adult Wings biological rep3  Adult Wings biological rep4  \\\n",
       "Gene Symbol                                                             \n",
       "128up                          181.05800                    204.36200   \n",
       "140up                           76.40320                     68.40710   \n",
       "18w                            752.14300                    754.71500   \n",
       "312                              8.25821                      5.44387   \n",
       "7B2                            128.78000                    183.37900   \n",
       "\n",
       "             Adult Female Spermatheca Mated (REDONE) biological rep1  \\\n",
       "Gene Symbol                                                            \n",
       "128up                                               380.878000         \n",
       "140up                                                85.908800         \n",
       "18w                                                 285.764000         \n",
       "312                                                   5.008850         \n",
       "7B2                                                   0.842605         \n",
       "\n",
       "             Adult Female Spermatheca Mated (REDONE) biological rep2  \\\n",
       "Gene Symbol                                                            \n",
       "128up                                                 112.6860         \n",
       "140up                                                 128.3340         \n",
       "18w                                                    66.5310         \n",
       "312                                                    35.0969         \n",
       "7B2                                                    16.0162         \n",
       "\n",
       "             Adult Female Spermatheca Mated (REDONE) biological rep3  \\\n",
       "Gene Symbol                                                            \n",
       "128up                                                294.49000         \n",
       "140up                                                 87.18940         \n",
       "18w                                                  248.04200         \n",
       "312                                                    2.39418         \n",
       "7B2                                                    2.64286         \n",
       "\n",
       "             Adult Female Spermatheca Mated (REDONE) biological rep4\"\\n  \n",
       "Gene Symbol                                                              \n",
       "128up                                               329.094000           \n",
       "140up                                               104.890000           \n",
       "18w                                                 264.250000           \n",
       "312                                                   2.270310           \n",
       "7B2                                                   0.579274           \n",
       "\n",
       "[5 rows x 136 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_expression.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The data frame has 18952 rows (measurements) and 136 columns (samples) so it is certainly high dimensional.These 136 columns represent 4 replicates each from 34 different tissue types."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will set numpy and pandas to display numbers to just two decimal places in this notebook - this won't affect the actual numbers, just their display, and you can change it if you prefer."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.set_printoptions(precision=2)\n",
    "pd.options.display.precision=2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Adult Hind Gut biological rep1</th>\n",
       "      <th>Adult Hind Gut biological rep2</th>\n",
       "      <th>Adult Hind Gut biological rep3</th>\n",
       "      <th>Adult Hind Gut biological rep4</th>\n",
       "      <th>Adult Mid Gut biological rep1</th>\n",
       "      <th>Adult Mid Gut biological rep2</th>\n",
       "      <th>Adult Mid Gut biological rep3</th>\n",
       "      <th>Adult Mid Gut biological rep4</th>\n",
       "      <th>Adult Accessory gland biological rep1</th>\n",
       "      <th>Adult Accessory gland biological rep2</th>\n",
       "      <th>...</th>\n",
       "      <th>Adult Thoracic Muscle biological rep3</th>\n",
       "      <th>Adult Thoracic Muscle biological rep4</th>\n",
       "      <th>Adult Wings biological rep1</th>\n",
       "      <th>Adult Wings biological rep2</th>\n",
       "      <th>Adult Wings biological rep3</th>\n",
       "      <th>Adult Wings biological rep4</th>\n",
       "      <th>Adult Female Spermatheca Mated (REDONE) biological rep1</th>\n",
       "      <th>Adult Female Spermatheca Mated (REDONE) biological rep2</th>\n",
       "      <th>Adult Female Spermatheca Mated (REDONE) biological rep3</th>\n",
       "      <th>Adult Female Spermatheca Mated (REDONE) biological rep4\"\\n</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>128up</th>\n",
       "      <td>214.51</td>\n",
       "      <td>246.60</td>\n",
       "      <td>216.58</td>\n",
       "      <td>221.67</td>\n",
       "      <td>264.60</td>\n",
       "      <td>281.02</td>\n",
       "      <td>300.13</td>\n",
       "      <td>255.19</td>\n",
       "      <td>427.69</td>\n",
       "      <td>467.31</td>\n",
       "      <td>...</td>\n",
       "      <td>113.94</td>\n",
       "      <td>363.60</td>\n",
       "      <td>196.98</td>\n",
       "      <td>262.51</td>\n",
       "      <td>181.06</td>\n",
       "      <td>204.36</td>\n",
       "      <td>380.88</td>\n",
       "      <td>112.69</td>\n",
       "      <td>294.49</td>\n",
       "      <td>329.09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140up</th>\n",
       "      <td>139.86</td>\n",
       "      <td>181.66</td>\n",
       "      <td>143.80</td>\n",
       "      <td>149.45</td>\n",
       "      <td>92.04</td>\n",
       "      <td>82.11</td>\n",
       "      <td>77.41</td>\n",
       "      <td>91.44</td>\n",
       "      <td>116.92</td>\n",
       "      <td>124.89</td>\n",
       "      <td>...</td>\n",
       "      <td>155.80</td>\n",
       "      <td>110.83</td>\n",
       "      <td>79.41</td>\n",
       "      <td>81.81</td>\n",
       "      <td>76.40</td>\n",
       "      <td>68.41</td>\n",
       "      <td>85.91</td>\n",
       "      <td>128.33</td>\n",
       "      <td>87.19</td>\n",
       "      <td>104.89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18w</th>\n",
       "      <td>241.72</td>\n",
       "      <td>247.03</td>\n",
       "      <td>245.75</td>\n",
       "      <td>262.19</td>\n",
       "      <td>54.11</td>\n",
       "      <td>60.05</td>\n",
       "      <td>60.70</td>\n",
       "      <td>54.94</td>\n",
       "      <td>33.03</td>\n",
       "      <td>39.50</td>\n",
       "      <td>...</td>\n",
       "      <td>100.41</td>\n",
       "      <td>294.09</td>\n",
       "      <td>772.39</td>\n",
       "      <td>700.45</td>\n",
       "      <td>752.14</td>\n",
       "      <td>754.72</td>\n",
       "      <td>285.76</td>\n",
       "      <td>66.53</td>\n",
       "      <td>248.04</td>\n",
       "      <td>264.25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>312</th>\n",
       "      <td>4.17</td>\n",
       "      <td>2.60</td>\n",
       "      <td>3.53</td>\n",
       "      <td>4.61</td>\n",
       "      <td>0.19</td>\n",
       "      <td>0.58</td>\n",
       "      <td>0.98</td>\n",
       "      <td>2.65</td>\n",
       "      <td>8.85</td>\n",
       "      <td>0.34</td>\n",
       "      <td>...</td>\n",
       "      <td>68.38</td>\n",
       "      <td>0.36</td>\n",
       "      <td>7.04</td>\n",
       "      <td>2.94</td>\n",
       "      <td>8.26</td>\n",
       "      <td>5.44</td>\n",
       "      <td>5.01</td>\n",
       "      <td>35.10</td>\n",
       "      <td>2.39</td>\n",
       "      <td>2.27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7B2</th>\n",
       "      <td>25.78</td>\n",
       "      <td>19.12</td>\n",
       "      <td>34.80</td>\n",
       "      <td>21.65</td>\n",
       "      <td>89.00</td>\n",
       "      <td>116.57</td>\n",
       "      <td>114.17</td>\n",
       "      <td>123.41</td>\n",
       "      <td>2.34</td>\n",
       "      <td>0.62</td>\n",
       "      <td>...</td>\n",
       "      <td>175.85</td>\n",
       "      <td>1.39</td>\n",
       "      <td>172.12</td>\n",
       "      <td>194.97</td>\n",
       "      <td>128.78</td>\n",
       "      <td>183.38</td>\n",
       "      <td>0.84</td>\n",
       "      <td>16.02</td>\n",
       "      <td>2.64</td>\n",
       "      <td>0.58</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 136 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Adult Hind Gut biological rep1  Adult Hind Gut biological rep2  \\\n",
       "Gene Symbol                                                                   \n",
       "128up                                214.51                          246.60   \n",
       "140up                                139.86                          181.66   \n",
       "18w                                  241.72                          247.03   \n",
       "312                                    4.17                            2.60   \n",
       "7B2                                   25.78                           19.12   \n",
       "\n",
       "             Adult Hind Gut biological rep3  Adult Hind Gut biological rep4  \\\n",
       "Gene Symbol                                                                   \n",
       "128up                                216.58                          221.67   \n",
       "140up                                143.80                          149.45   \n",
       "18w                                  245.75                          262.19   \n",
       "312                                    3.53                            4.61   \n",
       "7B2                                   34.80                           21.65   \n",
       "\n",
       "             Adult Mid Gut biological rep1  Adult Mid Gut biological rep2  \\\n",
       "Gene Symbol                                                                 \n",
       "128up                               264.60                         281.02   \n",
       "140up                                92.04                          82.11   \n",
       "18w                                  54.11                          60.05   \n",
       "312                                   0.19                           0.58   \n",
       "7B2                                  89.00                         116.57   \n",
       "\n",
       "             Adult Mid Gut biological rep3  Adult Mid Gut biological rep4  \\\n",
       "Gene Symbol                                                                 \n",
       "128up                               300.13                         255.19   \n",
       "140up                                77.41                          91.44   \n",
       "18w                                  60.70                          54.94   \n",
       "312                                   0.98                           2.65   \n",
       "7B2                                 114.17                         123.41   \n",
       "\n",
       "             Adult Accessory gland biological rep1  \\\n",
       "Gene Symbol                                          \n",
       "128up                                       427.69   \n",
       "140up                                       116.92   \n",
       "18w                                          33.03   \n",
       "312                                           8.85   \n",
       "7B2                                           2.34   \n",
       "\n",
       "             Adult Accessory gland biological rep2  ...  \\\n",
       "Gene Symbol                                         ...   \n",
       "128up                                       467.31  ...   \n",
       "140up                                       124.89  ...   \n",
       "18w                                          39.50  ...   \n",
       "312                                           0.34  ...   \n",
       "7B2                                           0.62  ...   \n",
       "\n",
       "             Adult Thoracic Muscle biological rep3  \\\n",
       "Gene Symbol                                          \n",
       "128up                                       113.94   \n",
       "140up                                       155.80   \n",
       "18w                                         100.41   \n",
       "312                                          68.38   \n",
       "7B2                                         175.85   \n",
       "\n",
       "             Adult Thoracic Muscle biological rep4  \\\n",
       "Gene Symbol                                          \n",
       "128up                                       363.60   \n",
       "140up                                       110.83   \n",
       "18w                                         294.09   \n",
       "312                                           0.36   \n",
       "7B2                                           1.39   \n",
       "\n",
       "             Adult Wings biological rep1  Adult Wings biological rep2  \\\n",
       "Gene Symbol                                                             \n",
       "128up                             196.98                       262.51   \n",
       "140up                              79.41                        81.81   \n",
       "18w                               772.39                       700.45   \n",
       "312                                 7.04                         2.94   \n",
       "7B2                               172.12                       194.97   \n",
       "\n",
       "             Adult Wings biological rep3  Adult Wings biological rep4  \\\n",
       "Gene Symbol                                                             \n",
       "128up                             181.06                       204.36   \n",
       "140up                              76.40                        68.41   \n",
       "18w                               752.14                       754.72   \n",
       "312                                 8.26                         5.44   \n",
       "7B2                               128.78                       183.38   \n",
       "\n",
       "             Adult Female Spermatheca Mated (REDONE) biological rep1  \\\n",
       "Gene Symbol                                                            \n",
       "128up                                                   380.88         \n",
       "140up                                                    85.91         \n",
       "18w                                                     285.76         \n",
       "312                                                       5.01         \n",
       "7B2                                                       0.84         \n",
       "\n",
       "             Adult Female Spermatheca Mated (REDONE) biological rep2  \\\n",
       "Gene Symbol                                                            \n",
       "128up                                                   112.69         \n",
       "140up                                                   128.33         \n",
       "18w                                                      66.53         \n",
       "312                                                      35.10         \n",
       "7B2                                                      16.02         \n",
       "\n",
       "             Adult Female Spermatheca Mated (REDONE) biological rep3  \\\n",
       "Gene Symbol                                                            \n",
       "128up                                                   294.49         \n",
       "140up                                                    87.19         \n",
       "18w                                                     248.04         \n",
       "312                                                       2.39         \n",
       "7B2                                                       2.64         \n",
       "\n",
       "             Adult Female Spermatheca Mated (REDONE) biological rep4\"\\n  \n",
       "Gene Symbol                                                              \n",
       "128up                                                   329.09           \n",
       "140up                                                   104.89           \n",
       "18w                                                     264.25           \n",
       "312                                                       2.27           \n",
       "7B2                                                       0.58           \n",
       "\n",
       "[5 rows x 136 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# The actual stored numbers have not changed\n",
    "raw_expression.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>It's common practice to take the log of expression values. Here is a visual motivation as to why this may be useful:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA8UAAAGDCAYAAAACrLI0AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjMsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+AADFEAAAgAElEQVR4nOzde5xlVX3n/c9XLopGBaVlkIY0kU4iOhG1BxidTFQyXNQEkkcSHJXWh5nOBROdmImtmQnxNg/mZoZnlAwJKKgRCYmBCIo9KDpORAElIKAPLaK0EAEbEG9g4+/5Y6/S09Wnqk5X16lTVfvzfr3O6+z922vvvfbu6lr1O2vtdVJVSJIkSZLURw+bdAUkSZIkSZoUk2JJkiRJUm+ZFEuSJEmSesukWJIkSZLUWybFkiRJkqTeMimWJEmSJPWWSbGkJS3JS5J8ZNL1kCT1S5Jbk/z8pOuxlCR5fZK/mnQ9pIUWv6dYkiRJ2l6SW4H/UFX/a9J1kTRe9hRLS0SS3Sddh121Eq5BkqSVIMluk66DtFyYFEsT1IZmvTbJdcC3k+yeZGOSLyW5P8mNSX5poPxXkjyzLb80SSU5tK3/hyR/P8N5Hp7kT5J8NcnXk/xFkr3attcmuXIqoU3yG0luSPKIJGvaOTYkuT3JHUleM3DcP0xyYZL3JPkm8PIkDxu4hm8kuSDJ41r5R7Sy30hyb5KrkuzXtr08yS3tur+c5CUD8U8OnPNZbb/72vuzBrZdkeRNSf5PO85Hkuy7UP9ekqR+au3on7e28Pa2/PCB7b/X2sjbW3tcSQ6Z4ViPTXJ2K/+1JG+eSmCTnJnkwoGyb01yeTrPSbKlDWG+u/0N8ZKBsu9q+1+a5NvAc+do//dN8sHWHm9N8r+TPKxte22r2/1JvpjkqBb/wyTvGTjnL7a/Ge5tbfCTB7bdmuR3k1zX2uz3J3nEgv2jSAvIpFiavBcDLwD2rqptwJeAnwUeC7wBeE+S/VvZjwPPacv/FrgF+LmB9Y/PcI63Aj8JHAYcAhwA/EHb9sfAg8B/SbIW+G/AS6vqewP7PxdYCxwNbMz2z1gdD1wI7A28F/ht4IRWrycC9wBvb2XXt+s6EHg88OvAd5M8CjgDOK6qHg08C7h2+kW05PqSVvbxwJ8BlyR5/ECxfw+8AngCsCfwuzPcE0mSRvX7wJF07ejTgMOB/wKQ5Fjgd4Cfp2tjf26GY0w5F9jWyj6drm39D23ba4CfaR8I/yxwCrC+fvS8478A9qVrx9cDZyX5qYFj/3vgLcCjgU8ye/v/GmALsArYD3g9UO14rwT+VWuTjwFunX4RSX4SeB/w6naMS4F/SLLnQLFfAY4FDgZ+Bnj5HPdGmgiTYmnyzqiq26rquwBV9TdVdXtV/aCq3g/cTNf4Qpf0TjW2Pwv8PwPrP8eQpDhJgP8I/Keq2lpV99Mlvie18/0AOJkumb0Y+KOq+ty0w7yhqr5dVdcD76RL5Kd8qqr+vtX3u8CvAb9fVVuq6gHgD4EXtZ7o79Mls4dU1UNVdU1VfbMd5wfAU5PsVVV3VNUNQ+7VC4Cbq+rdVbWtqt4HfAH4hYEy76yq/6/V5QK6PwQkSdoVLwHeWFV3VtVddB9av6xt+xW6tueGqvpO2zZUGx11HPDq1q7eCbyNH7XJ3wFeSveh73uA36qqLdMO81+r6oGq+jjdB8W/MrDtoqr6P61tf4BZ2n+6Nnl/4Mer6vtV9b9b8v0Q8HDg0CR7VNWtVfWlIZfzq8AlVbWpqr4P/AmwF90H21POaH/TbAX+AdtkLVEmxdLk3Ta4kuTkJNe2oUj3Ak+l+1QYuqT3Z5P8C2A34P3As5OsoeuB3aF3le7T20cC1wwc88MtDkBV3Qp8DFjDj3p1Z6rjV+h6gIfWH/hx4AMD57qJroHdD3g3cBlwfhti9ketwf02XeP668AdSS5J8tND6vHEdv5BX6H75HvKPw8sfwf4sSHHkSRpZ0xvfwbbwieyfVs4vV0c9OPAHnRt3VQ7+T/pRjcBUFWfoRsJFroPdwfd09rMYfWYfu652v8/BjYDH2mPL21s599M1/v7h8CdSc5PMniOKdvdk5aI34ZtspYhk2Jp8n44BXySHwf+km7Y0uOram/g83QN41RD9R26Xt1PtE99/xnYAHyyNUjT3Q18F3hKVe3dXo+tqh82TEmeD/xr4HK6RnK6AweWDwJuH1b/5ja6YdB7D7weUVVfa59Ev6GqDqX7JPmFdL3UVNVlVfXv6D61/kK7D9PdTvcHxaCDgK8NKStJ0kKZ3v4MtoV3AKsHtg22mdPdRteDu+9AG/mYqnrKVIEkp9L11N4O/N60/fdpjxwNqwds3ybP2v5X1f1V9Zqq+gm6EVe/M/XscFX9dVX9m3bNRTcMe7rt7kkbmXYgtslahkyKpaXlUXSNz10ASV5B11M86ON0SfPUUOkrpq1vpyXKfwm8LckT2nEPSHJMW94XOJvueab1wC+0JHnQf03yyCRPoXte9/2zXMNfAG9pCT5JViU5vi0/N8m/bBOKfJNu6NZDSfZrk3U8iu6PhW/R9S5Pdynwk0n+fbpJyX4VOBT44Cz1kSRpV72Pbu6NVa3d/AO64c3Q9ea+IsmTkzySHz2zu4OqugP4CPCnSR6TbnLKJyX5Ofjhc7pvphtC/TLg95JMH3L8hiR7tmeOXwj8zQznmqv9f2GSQ1oy+026dvehJD+V5HnpJhL7Hl1iPaxNvgB4QZKjkuxB94zyA8A/znT90lJlUiwtIVV1I/CnwKeArwP/Evg/04p9nG4CjU/MsD7Ma+mGSF2Zbpbo/wVMTcxxFt0zSJdW1TfoJvX4q2mTV3287X858CdV9ZFZzvXf6Z5N/kiS+4ErgSPatn9BNynXN+mGVX+c7o+Kh9E1prcDW+mej/7N6Qdu9XthK/sNuk/QX1hVd89SH0mSdtWbgauB64Drgc+2GFX1IboJID9G11Z+qu3zwAzHOpluIsgb6SajvBDYv8298R7grVX1T1V1M93kV+/Oj2a6/ue2z+10k1v+elV9YZZ6z9b+r23r32p1fkdVXUHXS306XU/zP9MN7X799ANX1Rfpkvf/t5X9BeAXqurBWeojLUn50WR2krS99qzyl4E92szYkiRpFu1riT4PPHwh284kzwHeU1Wr5yoraefYUyxJkiTtgiS/1IY070P3/O0/+GGytHyYFEuSJEm75tfo5gP5Et3zt78x2epI2hkOn5YkSZIk9ZY9xZIkSZKk3jIpliRJkiT11u6TrsBSse+++9aaNWsmXQ1J0gpxzTXX3F1VqyZdj+XMtlmStJBmaptNips1a9Zw9dVXT7oakqQVIslXJl2H5c62WZK0kGZqmx0+LUmSJEnqLZNiSZIkSVJvmRRLkiRJknrLpFiSJM0qya1Jrk9ybZKrW+xxSTYlubm979PiSXJGks1JrkvyjIHjrG/lb06yflLXI0nSIJNiSZI0iudW1WFVta6tbwQur6q1wOVtHeA4YG17bQDOhC6JBk4DjgAOB06bSqQlSZokk2JJkjQfxwPntuVzgRMG4udV50pg7yT7A8cAm6pqa1XdA2wCjl3sSkuSNJ1JsSRJmksBH0lyTZINLbZfVd0B0N6f0OIHALcN7LulxWaKS5I0UX5PsSRJmsuzq+r2JE8ANiX5wixlMyRWs8S337lLujcAHHTQQfOpqyRJO8WeYkmSNKuqur293wl8gO6Z4K+3YdG09ztb8S3AgQO7rwZunyU+/VxnVdW6qlq3atWqhb4USZJ2YFIsSZJmlORRSR49tQwcDXweuBiYmkF6PXBRW74YOLnNQn0kcF8bXn0ZcHSSfdoEW0e3mCRJE+XwaUmSNJv9gA8kge7vhr+uqg8nuQq4IMkpwFeBE1v5S4HnA5uB7wCvAKiqrUneBFzVyr2xqrYu3mVIkjScSbEkSZpRVd0CPG1I/BvAUUPiBZw6w7HOAc5Z6DpKkrQrHD4tSZIkSeotk+IFtmbjJazZeMmkqyFJkqRlwL8dpckzKZYkSZIk9ZZJsSRJkrSI7BmWlhYn2pIkSZIWmYmxtHTYUyxJkiRJ6i2TYkmSJGnCnHBLmhyHT0uSJEmLwKRXWppMiiVJkqQxMhmWljaHT0uSJEmSemtsSXGSRyT5TJJ/SnJDkje0+MFJPp3k5iTvT7Jniz+8rW9u29cMHOt1Lf7FJMcMxI9tsc1JNg7Eh55DkiRJWiw+JywtD+PsKX4AeF5VPQ04DDg2yZHAW4G3VdVa4B7glFb+FOCeqjoEeFsrR5JDgZOApwDHAu9IsluS3YC3A8cBhwIvbmWZ5RySJEmSJP3Q2JLi6nyrre7RXgU8D7iwxc8FTmjLx7d12vajkqTFz6+qB6rqy8Bm4PD22lxVt1TVg8D5wPFtn5nOIUmSJC1Z9i5Li2+szxS3Ht1rgTuBTcCXgHuralsrsgU4oC0fANwG0LbfBzx+MD5tn5nij5/lHNPrtyHJ1Umuvuuuu3blUiVJkiRJy9BYk+KqeqiqDgNW0/XsPnlYsfaeGbYtVHxY/c6qqnVVtW7VqlXDikiSJEmSVrBFmX26qu4FrgCOBPZOMvVVUKuB29vyFuBAgLb9scDWwfi0fWaK3z3LOSRJkqQlz2HU0uIZ5+zTq5Ls3Zb3An4euAn4GPCiVmw9cFFbvrit07Z/tKqqxU9qs1MfDKwFPgNcBaxtM03vSTcZ18Vtn5nOIUmSJI2VCa20vOw+d5F52x84t80S/TDggqr6YJIbgfOTvBn4HHB2K3828O4km+l6iE8CqKobklwA3AhsA06tqocAkrwSuAzYDTinqm5ox3rtDOeQJEmSJOmHxpYUV9V1wNOHxG+he754evx7wIkzHOstwFuGxC8FLh31HJIkSZIkDVqUZ4olSZIkSVqKTIolSZKkJcpnk6XxMymWJEmSJPXWOCfakiRJknrDXl1pebKnWJIkSZLUWybFkiRJkqTeMimWJEmSJPWWSbEkSZK0DKzZeInPLUtjYFIsSZIkSeotZ5+WJEmSljB7h6XxMimWJEmSdoFJq7S8OXxakiRJktRbJsWSJEmSpN4yKZYkSZKWEWehlhaWSbEkSZI0Tyan0vJnUixJkiQtQ/YYSwvDpFiSJEmS1FsmxZIkSZKk3jIpliRJkiT1lkmxJEmSJKm3TIolSZIkSb21+6QrIEmSJC03zvosrRz2FEuSJEnLmF/NJO0ak2JJkiRJUm+ZFEuSJElzsDdWWrlMiiVJkiRJvWVSLEmSJEnqLZNiSZIkSVJvmRRLkiRJknrLpFiSJEmS1Fu7T7oCkiRJ0nLhDNTSymNPsSRJkiSpt0yKJUmSJEm9ZVIsSZIkSeotk2JJkiRpBViz8RKfeZbmwaRYkiRJktRbJsWSJGlWSXZL8rkkH2zrByf5dJKbk7w/yZ4t/vC2vrltXzNwjNe1+BeTHDOZK5EkaUcmxZIkaS6vAm4aWH8r8LaqWgvcA5zS4qcA91TVIcDbWjmSHAqcBDwFOBZ4R5LdFqnukiTNyqRYkiTNKMlq4AXAX7X1AM8DLmxFzgVOaMvHt3Xa9qNa+eOB86vqgar6MrAZOHxxrkCSpNntPukKSJKkJe3Pgd8DHt3WHw/cW1Xb2voW4IC2fABwG0BVbUtyXyt/AHDlwDEH99lOkg3ABoCDDjpo4a5CmicnrpJWPnuKJUnSUEleCNxZVdcMhocUrTm2zbbP9sGqs6pqXVWtW7Vq1U7VV5Kk+RhbUpzkwCQfS3JTkhuSvKrF/zDJ15Jc217PH9hn6CQcSY5tsc1JNg7Ed3qiD0mSNLJnA7+Y5FbgfLph038O7J1karTZauD2trwFOBCgbX8ssHUwPmQfSZImapw9xduA11TVk4EjgVPbRBvQTc5xWHtdCjNPwtEm4ng7cBxwKPDigePs1EQfkiRpdFX1uqpaXVVr6Nroj1bVS4CPAS9qxdYDF7Xli9s6bftHq6pa/KT2ofXBwFrgM4t0GZIkzWpsSXFV3VFVn23L99PNWjn0+aFmpkk4Dgc2V9UtVfUg3SfVx89zog9JkrTrXgv8TpLNdM8Mn93iZwOPb/HfATYCVNUNwAXAjcCHgVOr6qFFr7UkSUMsyjPFbfjy04FPt9Ark1yX5Jwk+7TYDyfnaKYm4ZgpPvJEH8DURB/T67UhydVJrr7rrrt26RolSVrJquqKqnphW76lqg6vqkOq6sSqeqDFv9fWD2nbbxnY/y1V9aSq+qmq+tCkrkMaxXKfXGu5119abGNPipP8GPC3wKur6pvAmcCTgMOAO4A/nSo6ZPfZJueYz0Qf2weczEOSJEmSem2sSXGSPegS4vdW1d8BVNXXq+qhqvoB8Jf86HsKZ5qEY6b43ez8RB+SJElSb6zZeIk9x9Icxjn7dOieLbqpqv5sIL7/QLFfAj7flmeahOMqYG2baXpPuok+Lm4Td+zsRB+SJEnSnEwmpf7Yfe4i8/Zs4GXA9UmubbHX080efRjdcOZbgV+DbhKOJFOTcGxjYBKOJK8ELgN2A85pE3ZAN9HH+UneDHyO7Sf6eHeb6GMrXSItSZIkSdJ2xpYUV9UnGf5s76Wz7PMW4C1D4pcO269N4HH4kPj3gBN3pr6SJEnqt5XUM7ySrkUat0WZfVqSJEmSpKXIpFiSJEmS1FsmxZIkSZKk3jIpliRJkiT1lkmxJEmSJKm3TIolSZIkSb1lUixJkiRJ6q2xfU+xJEmStBz4nb5Sv9lTLEmSJEnqLXuKJUmS1Et96iGeutZbT3/BhGsiLT32FEuSJEmSesukWJIkSZLUWybFkiRJkqTeMimWJEmSJPWWSbEkSZIkqbdMiiVJkqSeWLPxkl7Nui2NwqRYkiRJktRbJsWSJEmSpN4yKZYkSZIk9ZZJsSRJknrB52klDWNSLEmSJEnqLZNiSZIkSVJv7T7pCkiSJEmLySHUkgbZUyxJkiRJ6i2TYkmSJElSb5kUS5IkSZJ6y6RYkiRJK57PEUuaiUmxJEmS1FN+WCA5+7QkSZLUOybD0o/YUyxJkiRJ6i17iiVJkrRi2SM6t6l7dOvpL5hwTaTJsKdYkiRJktRbJsWSJEmSpN4yKZYkSZIk9ZZJsSRJkiSpt0yKJUmSJEm9ZVIsSZIkSeotk2JJkiRJUm+ZFEuSJOmH1my8xO/2ldQrcybFSf4oyWOS7JHk8iR3J3npYlROkiQtDNtzSZKGG6Wn+Oiq+ibwQmAL8JPAf55rpyQHJvlYkpuS3JDkVS3+uCSbktzc3vdp8SQ5I8nmJNclecbAsda38jcnWT8Qf2aS69s+ZyTJbOeQJKnH5tWeS8uNPd3z571TX42SFO/R3p8PvK+qto547G3Aa6rqycCRwKlJDgU2ApdX1Vrg8rYOcBywtr02AGdCl+ACpwFHAIcDpw0kuWe2slP7HdviM51DkqS+mm97rhXIxEeSfmT3Ecr8Q5IvAN8FfjPJKuB7c+1UVXcAd7Tl+5PcBBwAHA88pxU7F7gCeG2Ln1dVBVyZZO8k+7eym6Ya7ySbgGOTXAE8pqo+1eLnAScAH5rlHJIk9dW82nNpuTLxlzSqOXuKq2oj8K+BdVX1feA7dEnnyJKsAZ4OfBrYryXMU4nzE1qxA4DbBnbb0mKzxbcMiTPLOSRJ6qWFaM8lSVqJ5uwpTvJI4FTgILqhyk8Efgr44CgnSPJjwN8Cr66qb7bHfocWHRKrecRHlmQD3TVx0EEH7cyukiQtK7vanmvlmupRvfX0FwyNT5m+XZJWilGeKX4n8CDwrLa+BXjzKAdPsgddQvzeqvq7Fv56GxZNe79z4LgHDuy+Grh9jvjqIfHZzrGdqjqrqtZV1bpVq1aNckmSJC1X827PpSlOxCRpJRolKX5SVf0R8H2Aqvouw3tpt9Nmgj4buKmq/mxg08XA1AzS64GLBuInt1mojwTua0OfLwOOTrJPm2DraOCytu3+JEe2c5087VjDziFJUl/Nqz3XyjU9wZ0r4TUZlrRSjTLR1oNJ9qINTU7yJOCBEfZ7NvAy4Pok17bY64HTgQuSnAJ8FTixbbuUbkbMzXTPOb0CoKq2JnkTcFUr98aBGTN/A3gXsBfdBFsfavGZziFJUl/Ntz3XCmJiq1HMNKReWqlGSYpPAz4MHJjkvXTJ7svn2qmqPsnMn0AfNaR80T3rNOxY5wDnDIlfDTx1SPwbw84hSVKPzas9lyRppZszKa6qTUk+S/ddwwFeVVV3j71mkiRpwcy3PU/yCOATwMPp/m64sKpOS3IwcD7wOOCzwMuq6sEkDwfOA54JfAP41aq6tR3rdcApwEPAb1fVZQt8mVok9iRKWknmfKY4yS8B26rqkqr6ILAtyQnjr5okSVoou9CePwA8r6qeBhwGHNvm/ngr8LaqWgvcQ5fs0t7vqapDgLe1ciQ5FDgJeApwLPCOJLst3BWq75wETNJ8jTLR1mlVdd/USlXdSzcES5IkLR/zas+r8622ukd7FfA84MIWPxeYSrCPb+u07Ue1CTGPB86vqgeq6st0c4gcvmuXJEnSrhvlmeJhifMo+0mSpKVj3u1569G9BjgEeDvwJeDeqtrWimwBDmjLBwC3AVTVtiT3AY9v8SsHDju4j8bMHlRJmtkojeHVSf6MrhEs4LfoGkZJkrR8zLs9r6qHgMOS7A18AHjysGLtfdgkmzVLfDtJNgAbAA466KBRqqdZrORk2OeaJS2UUZLi3wL+K/B+ugbtI8wwS7QkSVqydrk9r6p7k1xBN1nX3kl2b73Fq4HbW7EtwIHAliS7A48Ftg7EpwzuM3iOs4CzANatW7dD0ixp8Uz/UMUPILRSjTL79LeBjYtQF0mSNCbzbc+TrAK+3xLivYCfp5s862PAi+hmoF4PXNR2ubitf6pt/2hVVZKLgb9uvdVPBNYCn9m1q5IkadfNmRQn+Ungd4E1g+Wr6nnjq5YkSVpIu9Ce7w+c254rfhhwQVV9MMmNwPlJ3gx8Dji7lT8beHeSzXQ9xCe189yQ5ALgRmAbcGobli1J0kSNMnz6b4C/AP6K7nsFJUnS8jOv9ryqrgOePiR+C0Nmj66q7wEnznCstwBvGfXckiQthlGS4m1VdebYayJJksbJ9lwrxkqeQEzS4hslKf6HJL9JN9vkA1PBqto6tlpJkqSFZnveI31KGvt0rZLGY5SkeH17/88DsQJ+YuGrI0mSxsT2XAvOr0WStBKMMvv0wYtREUmSND625/0wqV5Tk2NJy9nD5iqQ5JFJ/kuSs9r62iQvHH/VJEnSQrE9lyRpuDmTYuCdwIPAs9r6FuDNY6uRJEkaB9tzSZKGGCUpflJV/RHwfYCq+i6QsdZKkiQtNNvzFW4lTzi1ZuMlK/r6JE3WKEnxg0n2opuMgyRPYmDWSkmStCzYnmvRDCawJrSSlrpRZp8+DfgwcGCS9wLPBl4+zkpJkqQFZ3uusdvV5Hf6hF0m00vL4L+Hk6ppJZk1KU4S4AvALwNH0g2zelVV3b0IdZMkSQvA9lzLjcmwpMU0a1JcVZXk76vqmYC/nSRJWoZsz7WU+PVNkpaaUZ4pvjLJvxp7TSRJ0jjZnmtJs3dY0qSM8kzxc4FfS/IV4Nt0Q66qqn5mrDWTJEkLyfZci8okV9JyMUpSfNzYayFJksbN9lxL0s4kzybaksZhrom2HgZcUlVPXaT6SJKkBWZ7rqXAhFbSUjXrM8VV9QPgn5IctEj1kSRJC8z2XJKkmY0yfHp/4IYkn6F7BgmAqvrFsdVKkiQtNNtzLSn2HEtaKkZJit8w9lpIkqRxsz1foUwuJWnXzJkUV9XHF6MikiRpfGzPJUkabs6kOMn9QLXVPYE9gG9X1WPGWTFJkrRwbM8lSRpulJ7iRw+uJzkBOHxsNZIkSQvO9nzlcdi0JC2MWWefHqaq/h543hjqIkmSFont+cJZs/ESE1T1jj/3WklGGT79ywOrDwPW8aPhV5IkaRmwPR8PkwJJWv5GmX36FwaWtwG3AsePpTaSJGlcbM8lSRpilGeKX7EYFZEkSeNje76w7CGWpJVjzmeKk5ybZO+B9X2SnDPeakmSpIVke75y+CynlhJ/HrUSjDLR1s9U1b1TK1V1D/D08VVJkiSNge25JElDjJIUPyzJPlMrSR7HaM8iS5KkpcP2XJKkIUZpDP8U+MckF9LNUvkrwFvGWitJkrTQbM8lSRpilIm2zktyNd13GQb45aq6cew1kyRJC8b2fGHM9uzk1LZbT3/BYlVHkrQARvme4iOBG6rqf7T1Ryc5oqo+PfbaSZKkBWF7Lmmc/FBIy9kow6fPBJ4xsP7tITFJkrS02Z7vgp2ZXXd6WZMESVraRploK1VVUytV9QNG62E+J8mdST4/EPvDJF9Lcm17PX9g2+uSbE7yxSTHDMSPbbHNSTYOxA9O8ukkNyd5f5I9W/zhbX1z275mhGuUJGmlm1d7LknSSjdKUnxLkt9Oskd7vQq4ZYT93gUcOyT+tqo6rL0uBUhyKHAS8JS2zzuS7JZkN+DtwHHAocCLW1mAt7ZjrQXuAU5p8VOAe6rqEOBtrZwkSX033/ZckqQVbZSk+NeBZwFfa68jgA1z7VRVnwC2jliP44Hzq+qBqvoysBk4vL02V9UtVfUgcD5wfJLQTRRyYdv/XOCEgWOd25YvBI5q5SVJ6rN5teeSJK10o8w+fSddL+5CeWWSk4GrgddU1T3AAcCVA2W2tBjAbdPiRwCPB+6tqm1Dyh8wtU9VbUtyXyt/9wJegyRJy8oY2nNJklaEOXuKk6xO8oH2fPDXk/xtktXzPN+ZwJOAw4A76L4zEbqvhpiu5hGf7Vg7SLIhydVJrr7rrrtmq7ckScvaArfnmoA1Gy/ZqQm/JEmjGWX49DuBi4En0vXC/kOL7bSq+npVPdQm9/hLuuHR0PX0HjhQdDVw+yzxu4G9k+w+Lb7dsdr2xzLDMO6qOquq1lXVulWrVs3nkiRJWi4WrD2XJGklGSUpXlVV76yqbe31LmBeGWSS/QdWfwmYmpn6YuCkNnP0wcBa4DPAVcDaNtP0nnTDvi5us2d+DHhR2389cNHAsda35RcBHx2cbVOSpJ5asPZc82MvryQtTaN8FcPdSV4KvK+tvxj4xlw7JXkf8Bxg3yRbgNOA5yQ5jG44863AryJx2r8AABlGSURBVAFU1Q1JLgBuBLYBp1bVQ+04rwQuA3YDzqmqG9opXgucn+TNwOeAs1v8bODdSTbT9RD7/JQkSfNszzU5JtFajqZ+bv1+bi0noyTF/zfwP+i+3qiAf2yxWVXVi4eEzx4Smyr/FuAtQ+KXApcOid/Cj4ZfD8a/B5w4V/0kSeqZebXnfWdiKkkr3yizT38V+MVFqIskSRoT2/PJMbGWpKVtlGeKJUmSJElakUyKJUmSJEm9ZVIsSZK0RPhdxJK0+OZ8pjjJ3sDJwJrB8lX12+OrliRJWki255IW0+As1M5IraVulNmnLwWuBK4HfjDe6kiSpDGxPV8CTA4kaekZJSl+RFX9zthrIkmSxsn2fAmaKUl2CLVWCn+WtRyM8kzxu5P8xyT7J3nc1GvsNZMkSQvJ9lySpCFG6Sl+EPhj4PeBarECfmJclZIkSQvO9lySpCFGSYp/Bzikqu4ed2UkSdLY2J5LkjTEKMOnbwC+M+6KSJKksbI9lyRpiFF6ih8Crk3yMeCBqaBf4SBJ0rJie74TFntyICcjUh84+7qWqlGS4r9vL0mStHzZnkuSNMScSXFVnbsYFZEkSeNje7602DMsSUvHnM8UJ/lyklumvxajcpIkaWHMtz1PcmCSjyW5KckNSV7V4o9LsinJze19nxZPkjOSbE5yXZJnDBxrfSt/c5L147taSZJGN8rw6XUDy48ATgT8XkNJkpaX+bbn24DXVNVnkzwauCbJJuDlwOVVdXqSjcBG4LXAccDa9joCOBM4on0n8mmtHtWOc3FV3bMgVydJ0jzN2VNcVd8YeH2tqv4ceN4i1E2SJC2Q+bbnVXVHVX22Ld8P3AQcABwPTA3JPhc4oS0fD5xXnSuBvZPsDxwDbKqqrS0R3gQcu5DXKEnSfMzZUzw47IkuiV4HPHpsNZIkSQtuIdrzJGuApwOfBvarqjugS5yTPKEVOwC4bWC3LS02U3z6OTYAGwAOOuignameJEnzMsrw6T8dWN4G3Ar8ylhqI0mSxmWX2vMkPwb8LfDqqvpmkhmLDonVLPHtA1VnAWcBrFu3boftkiQttFFmn37uYlREkiSNz66050n2oEuI31tVf9fCX0+yf+sl3h+4s8W3AAcO7L4auL3FnzMtfsV86yRJ0kIZZfj0w4H/C1gzWL6q3ji+akmSpIU03/Y8XZfw2cBNVfVnA5suBtYDp7f3iwbir0xyPt1EW/e1xPky4L9NzVINHA28bleva6H5VUmS1D+jDJ++CLgPuAZ4YLzVkSRJYzLf9vzZwMuA65Nc22Kvp0uGL0hyCvBVutmsAS4Fng9sBr4DvAKgqrYmeRNwVSv3xqraOv/LkSRpYYySFK+uKmeHlCRpeZtXe15Vn2T488AARw0pX8CpMxzrHOCcna2DJEnjNOdXMgH/mORfjr0mkiRpnGzPJUkaYpSe4n8DvDzJl+mGW4Xug+CfGWvNJEnSQrI9l7QkTD27f+vpL5hwTaTOKEnxcWOvhSRJGjfbc0lLismxlopRvpLpK4tREUmSND6255IkDTfKM8WSJEmSNBZrNl7i16FpokyKJUmSJEm9ZVIsSZIkSeotk2JJkiRJUm+ZFEuSJEmSesukWJIkSZLUWybFkiRJkqTeMimWJEmSJPWWSbEkSZIkqbdMiiVJkiRJvWVSLEmSJGni1my8ZNJVUE+ZFEuSJElaEtZsvMTkWIvOpFiSJEmS1FsmxZIkSZKk3hpbUpzknCR3Jvn8QOxxSTYlubm979PiSXJGks1JrkvyjIF91rfyNydZPxB/ZpLr2z5nJMls55AkSZIkabpx9hS/Czh2WmwjcHlVrQUub+sAxwFr22sDcCZ0CS5wGnAEcDhw2kCSe2YrO7XfsXOcQ5IkSdIy4LPFWkxjS4qr6hPA1mnh44Fz2/K5wAkD8fOqcyWwd5L9gWOATVW1taruATYBx7Ztj6mqT1VVAedNO9awc0iSJEmStJ3FfqZ4v6q6A6C9P6HFDwBuGyi3pcVmi28ZEp/tHDtIsiHJ1Umuvuuuu+Z9UZIkSZKk5WmpTLSVIbGaR3ynVNVZVbWuqtatWrVqZ3eXJEmSJC1zi50Uf70Nfaa939niW4ADB8qtBm6fI756SHy2c0iSJM3I5xclqZ8WOym+GJiaQXo9cNFA/OQ2C/WRwH1t6PNlwNFJ9mkTbB0NXNa23Z/kyDbr9MnTjjXsHJIkSZIkbWf3cR04yfuA5wD7JtlCN4v06cAFSU4Bvgqc2IpfCjwf2Ax8B3gFQFVtTfIm4KpW7o1VNTV512/QzXC9F/Ch9mKWc0iSJEmStJ2xJcVV9eIZNh01pGwBp85wnHOAc4bErwaeOiT+jWHnkCRJkiRpuqUy0ZYkSZIkSYvOpFiSJEnSkrRm4yVOgqexMymWJEmStKSZHGucTIolSZIkSb1lUixJkiRJ6i2TYkmSJElSb5kUS5IkSVoWfLZY42BSLEmSJEnqLZNiSZIkSVJvmRRLkiRJknrLpFiSJEmS1FsmxZIkSZKWFSfc0kIyKZYkSZIk9ZZJsSRJkiSpt0yKJUmSJEm9ZVIsSZIkaVnyuWItBJNiSZIkSVJvmRRLkiRJWraciVq7yqRYkiRJktRbJsWSJEmSpN7afdIVkCRJmiSHXUpSv9lTLEmSJEnqLZNiSZIkScueE25pvkyKJUmSJEm9ZVIsSZIkSeotk2JJkiRJUm+ZFEuSJEmSesukWJIkzSjJOUnuTPL5gdjjkmxKcnN736fFk+SMJJuTXJfkGQP7rG/lb06yfhLXIqkfnHBLO8ukWJIkzeZdwLHTYhuBy6tqLXB5Wwc4DljbXhuAM6FLooHTgCOAw4HTphJpSZImzaRYkiTNqKo+AWydFj4eOLctnwucMBA/rzpXAnsn2R84BthUVVur6h5gEzsm2pI0FvYcay67T7oCkiRp2dmvqu4AqKo7kjyhxQ8Abhsot6XFZorvIMkGul5mDjrooAWutqQ+MRHWqOwpliRJCyVDYjVLfMdg1VlVta6q1q1atWpBKydJ0jAmxZIkaWd9vQ2Lpr3f2eJbgAMHyq0Gbp8lLknSxJkUS5KknXUxMDWD9HrgooH4yW0W6iOB+9ow68uAo5Ps0ybYOrrFJEmaOJ8pliRJM0ryPuA5wL5JttDNIn06cEGSU4CvAie24pcCzwc2A98BXgFQVVuTvAm4qpV7Y1VNn7xLkqSJMCmWJEkzqqoXz7DpqCFlCzh1huOcA5yzgFWTJGlBOHxakiRJktRbJsWSJEmSesnvMBaYFEuSJEnqAZNfzcRniiVJkiT1gomxhrGnWJIkSZLUWxNJipPcmuT6JNcmubrFHpdkU5Kb2/s+LZ4kZyTZnOS6JM8YOM76Vv7mJOsH4s9sx9/c9s3iX6UkSZKk5WCmZ4vtWe6HSfYUP7eqDquqdW19I3B5Va0FLm/rAMcBa9trA3AmdEk03XclHgEcDpw2lUi3MhsG9jt2/JcjSZIkSU7gtdwspeHTxwPntuVzgRMG4udV50pg7yT7A8cAm6pqa1XdA2wCjm3bHlNVn2rfl3jewLEkSZIkad7sVV55JpUUF/CRJNck2dBi+1XVHQDt/QktfgBw28C+W1pstviWIfEdJNmQ5OokV9911127eEmSJEmSpOVmUrNPP7uqbk/yBGBTki/MUnbY88A1j/iOwaqzgLMA1q1bN7SMJEmSpH6Y6u299fQXzBqbLa7lZyJJcVXd3t7vTPIBumeCv55k/6q6ow2BvrMV3wIcOLD7auD2Fn/OtPgVLb56SHlJkiRJmpf5DI+evo8J9NK06MOnkzwqyaOnloGjgc8DFwNTM0ivBy5qyxcDJ7dZqI8E7mvDqy8Djk6yT5tg62jgsrbt/iRHtlmnTx44liRJkiRJPzSJnuL9gA+0b0naHfjrqvpwkquAC5KcAnwVOLGVvxR4PrAZ+A7wCoCq2prkTcBVrdwbq2prW/4N4F3AXsCH2kuSJOmHnBRH0kIY/F3i75XladGT4qq6BXjakPg3gKOGxAs4dYZjnQOcMyR+NfDUXa7sLvAZA0mSJGl5Mrntl6X0lUySJEmSJC0qk2JJkiRJUm+ZFEuSJEmSesukWJIkSZIWwZqNl/i88hI0ke8pliRJkqS+GkyMnZh38uwpliRJkiT1lkmxJEmSJKm3TIolSZIkSb1lUixJkiRJ6i2TYkmSJElSb5kUS5IkSZJ6y6RYkiRJkibE7y6ePJNiSZIkSVJvmRRLkiRJknrLpFiSJEmS1FsmxZIkSZI0YT5bPDkmxZIkSZKk3jIpliRJkiT1lkmxJEmSJKm3TIolSZIkSb1lUixJkiRJS4QTbi0+k2JJkiRJUm+ZFEuSJEmSesukWJIkSZLUWybFkiRJkrTE+Gzx4tl90hVY6aZ+kG89/QUTrokkSQL8I1OStB17iiVJkiRJvWVSLEmSJEnqLZNiSZIkSVqifLZ4/EyKJUmSJGmJMzEeH5NiSZIkSVoG7DUeD5NiSZIkSVJv+ZVMkiRJkrSMTO8t9utfd409xYvEoQ6SJEmStPSYFEuSJEnSMmYH3K4xKZYkSZIk9ZZJsSRJkiStAPYYz49J8SLzB1WSJEnSOJlv7Bxnn5YkSb3gH4mSpGHsKZ4QG2ZJkiRJ4+II1dGt2KQ4ybFJvphkc5KNk66PJEl9Z9ssSYvP5HhuK3L4dJLdgLcD/w7YAlyV5OKqunGyNdueX7otSeqL5dI2S9JKNZV7mHPsaEUmxcDhwOaqugUgyfnA8cCSbnhn+gTHH1xJ0gqwLNtmSVpp7Jjb0UpNig8AbhtY3wIcMaG67LJJDHfwP4ckaYFNrG122KAkzWxXf0dOzxtm65Feqr3VKzUpzpBY7VAo2QBsaKvfSvLFBTj3vsDdC3CcicpbJ3r6FXEPJ8j7t2u8f7vG+/cjPz7pCiwxts1Li/dke96PHXlPduQ92dG+eevwezJbPjHBXGNo27xSk+ItwIED66uB26cXqqqzgLMW8sRJrq6qdQt5zL7xHu4a79+u8f7tGu+fZmHbvIR4T7bn/diR92RH3pMdrZR7slJnn74KWJvk4CR7AicBF0+4TpIk9ZltsyRpSVqRPcVVtS3JK4HLgN2Ac6rqhglXS5Kk3rJtliQtVSsyKQaoqkuBSydw6gUd8tVT3sNd4/3bNd6/XeP904xsm5cU78n2vB878p7syHuyoxVxT1K1wxwXkiRJkiT1wkp9pliSJEmSpDmZFC+gJMcm+WKSzUk2Tro+k5TknCR3Jvn8QOxxSTYlubm979PiSXJGu2/XJXnGwD7rW/mbk6wfiD8zyfVtnzOSDPuqj2UryYFJPpbkpiQ3JHlVi3sPR5DkEUk+k+Sf2v17Q4sfnOTT7V68v032Q5KHt/XNbfuagWO9rsW/mOSYgfiK//+eZLckn0vywbbu/dOy4s/Z9mZqW7Tj77u+S7J3kguTfKH9vPzrSddp0pL8p/b/5vNJ3pfkEZOu02Lbmb/vlxuT4gWSZDfg7cBxwKHAi5McOtlaTdS7gGOnxTYCl1fVWuDytg7dPVvbXhuAM6H7TwacBhwBHA6cNvAf7cxWdmq/6eda7rYBr6mqJwNHAqe2nyfv4WgeAJ5XVU8DDgOOTXIk8Fbgbe3+3QOc0sqfAtxTVYcAb2vlaPf8JOApdPfnHe0Pp778f38VcNPAuvdPy4Y/Z0PN1LZox993ffffgQ9X1U8DT6Pn9ybJAcBvA+uq6ql0kwWeNNlaTcS7GP3v+2XFpHjhHA5srqpbqupB4Hzg+AnXaWKq6hPA1mnh44Fz2/K5wAkD8fOqcyWwd5L9gWOATVW1taruATbRJTf7A4+pqk9V91D8eQPHWhGq6o6q+mxbvp+uMToA7+FI2n34Vlvdo70KeB5wYYtPv39T9/VC4KjWc348cH5VPVBVXwY20/1fX/H/35OsBl4A/FVbD94/LS/+nE0zS9vSa9N/3/VdkscA/xY4G6CqHqyqeydbqyVhd2CvJLsDj2TI96yvdDv59/2yYlK8cA4AbhtY34INzXT7VdUd0DXMwBNafKZ7N1t8y5D4itSGoj4d+DTew5G1HslrgTvpPgz4EnBvVW1rRQav+Yf3qW2/D3g8O39fV5I/B34P+EFbfzzePy0v/pzNYlrb0nfTf9/13U8AdwHvbEPK/yrJoyZdqUmqqq8BfwJ8FbgDuK+qPjLZWi0ZM/1tuqyYFC+cYc9jOrX3aGa6dzsbX3GS/Bjwt8Crq+qbsxUdEuv1Payqh6rqMGA1XY/Rk4cVa+/evwFJXgjcWVXXDIaHFPX+aSnz52wGO9G2rHgz/L7ru92BZwBnVtXTgW+zTIfELpT26NnxwMHAE4FHJXnpZGulhWRSvHC2AAcOrK+mh8Mq5vD1NmyX9n5ni89072aLrx4SX1GS7EH3R8t7q+rvWth7uJPakK8r6J6f27sNe4Ltr/mH96ltfyzd8KCdva8rxbOBX0xyK92Q0+fR9aR4/7Sc+HM2xAxtS5/t8PsuyXsmW6WJ2wJsqaqpUQQX0iXJffbzwJer6q6q+j7wd8CzJlynpWKmv02XFZPihXMVsLbNzron3cP3F0+4TkvNxcDU7MfrgYsG4iencyTdkJQ7gMuAo5Ps0z6hOxq4rG27P8mR7bnFkweOtSK06zobuKmq/mxgk/dwBElWJdm7Le9F15jdBHwMeFErNv3+Td3XFwEfbc9aXwyclG525YPpJiT7DCv8/3tVva6qVlfVGrpr+2hVvQTvn5YXf86mmaVt6a0Zft/1ugewqv4ZuC3JT7XQUcCNE6zSUvBV4Mgkj2z/j46i55OPDZjpb9NlZfe5i2gUVbUtySvpkpDdgHOq6oYJV2tikrwPeA6wb5ItdDMgnw5ckOQUul8uJ7bilwLPp5uE5zvAKwCqamuSN9H9YQPwxqqaerj/N+hmwNsL+FB7rSTPBl4GXN+eiwV4Pd7DUe0PnNtmn30YcEFVfTDJjcD5Sd4MfI42iUh7f3eSzXQ9nCcBVNUNSS6g+2NgG3BqVT0E0NP/76/F+6dlwnZ5qKFtS1VdOsE6aWn6LeC97QOlW2h/V/RVVX06yYXAZ+nas88BZ022VotvJ/++X1bSfZgvSZIkSVL/OHxakiRJktRbJsWSJEmSpN4yKZYkSZIk9ZZJsSRJkiSpt0yKJUmSJEm9ZVIsaU5J/nFC531XkhfNXVKSpH6xbZYWjkmxtIK17+ndZVX1rIU4jiRJfWfbLC09JsXSEpHkpUk+k+TaJP8zyW5JfjzJzUn2TfKwJP87ydFJ1iT5QpJzk1yX5MIkj2zHuTXJHyT5JHBikicl+XCSa9r+P93KnZjk80n+KcknWuwpA3W4LsnaFv9We0+SP277XZ/kV1v8OUmuaPX4QpL3Jsm063tyks8MrK9Jcl1b/oMkV7XjnjV934Hr2rctr0tyRVt+VJJz2v6fS3L8Qv/bSJL6ybbZtln9YFIsLQFJngz8KvDsqjoMeAh4SVV9BXgr8BfAa4Abq+ojbbefAs6qqp8Bvgn85sAhv1dV/6aqzgfOAn6rqp4J/C7wjlbmD4BjquppwC+22K8D/73VYR2wZVpVfxk4DHga8PPAHyfZv217OvBq4FDgJ4BnD+5YVTcBeyb5iRb6VeCCtvw/qupfVdVTgb2AF45y35rfB/7/du4m1KoqjMP4848EIZsIBkFdJSdB0AcmUkhhQZOGVhIJERU0cSZ9YVLckTSqSUQNKjK5BRXRxKJLXxL3OrleLYnoA2xQ4ECKCCN9Hex18HTuueIZeWM/v8lZe6+9Fu8+g/Oetfbaa7aqNgPbWkxXTNBekqQlzM3mZvWHg2JpZbgb2AQcTrLQjq8DqKrXgSvpkuLuoTYnqupQK78NbB2qmwFIsga4HXiv9fsqMEiUh4A3kjwODJZyfQM8m+QpYH1V/T0S51bgQFWdqarfgS+Aza1uvqp+raqzwAKwYcx9vgs80Mo7BnEC25LMJTkK3AXcMP5rGuse4Ol2f58Dq4GpCdpLkjSOudncrJ64/FIHIAmAAG9W1TNLKrqlV9e0wzXAn61cI5cOH//VPi8DTrXZ5f9eXPVEki3AvcBCkpur6p0kc+3cwSSPVdXsSJzLOT1UPsP435cZuj8B73ch1A9JVtPNkN9aVSeSPE+XPEf9y/mJvOH6ANur6vsLxCZJ0qTMzeZm9YRPiqWV4TPgviRXASRZm2R9q9sH7KdbUvXaUJupJLe18oPA16OdVtUfwM9J7m/9JslNrbyxquaqai9wEri2LZ/6qapeBj4Cbhzp8ktgR3unah1wBzDPRaqqH+mS8nOcn4keJNGTbfZ8uR0tf6GbsQfYPnT+ILBr8K5TklsuNh5Jki7A3GxuVk84KJZWgKr6DtgDfNI2uPgUuDrJnXRLoPZV1X7gnySPtGbHgYfb9WuBV5bp/iHg0SRHgG+BwWYXL7YNOY7RJdQjdMumjrXlTtcDb4309QGw2K6dBZ6sqt8mvN0ZYCftnaWqOkX3h+Io8CFweJl2LwAvJfmKLnkPTAOrgMV2L9MTxiNJ0hLmZnOz+iNVo6s8JK10STYAH7fNLyRJ0iVmbpb+v3xSLEmSJEnqLZ8US5IkSZJ6yyfFkiRJkqTeclAsSZIkSeotB8WSJEmSpN5yUCxJkiRJ6i0HxZIkSZKk3nJQLEmSJEnqrXMGfTh31R1TIQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 1152x432 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "log_expression = np.log(raw_expression + 1)\n",
    "\n",
    "fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(16, 6))\n",
    "ax1.hist(raw_expression.values.flatten(), bins=200)\n",
    "ax1.set_title('raw expression')\n",
    "ax1.set_xlabel('expression value')\n",
    "ax1.set_ylabel('num occurrences')\n",
    "ax2.hist(log_expression.values.flatten(), bins=200)\n",
    "ax2.set_title('log expression')\n",
    "ax2.set_xlabel('expression value')\n",
    "ax2.set_ylabel('num occurrences')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "From here on, we will use 'log_expression' as our sample data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "log_expression = np.log(raw_expression + 1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "The following code snippet removes the replicate name from each sample, so we can use these labels as categories for plotting later."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "tissues_list = [re.match('(.+?)(( biological)? rep\\d+)', c).group(1)\n",
    "                     for c in raw_expression.columns]\n",
    "tissues = pd.Series(tissues_list, index = raw_expression.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "136"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(tissues)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br><br>\n",
    "### Task 1 - Building a correlation matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Task (1a)**\n",
    "\n",
    "The [FlyAtlas](http://flyatlas.org) dataset contains four biological replicates for each tissue. Combine the biological replicates by calculating the mean expression value for each gene in each tissue."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ~~ GRADED CELL (1 mark) - complete this cell ~~\n",
    "\n",
    "def average_by_tissue(expression, tissues):\n",
    "    '''\n",
    "    Given a DataFrame of gene expression data, \n",
    "    and a list, array or Series of tissues corresponding to the columns of the dataframe,\n",
    "    average over the expression values in each gene for each tissue type and\n",
    "    return the resulting dataframe. \n",
    "    The columns of the new dataframe should correspond to the provided tissues.\n",
    "    '''\n",
    "    # Since Python set is unordered, additional unique_tissues list is added to preserve order\n",
    "    unique_tissues = []\n",
    "    unique_tissues_set = set()\n",
    "    for tissue in tissues:\n",
    "        if tissue not in unique_tissues_set:\n",
    "            unique_tissues.append(tissue)\n",
    "            unique_tissues_set.add(tissue)\n",
    "\n",
    "    # Store every average values inside a dictionary\n",
    "    averaged_expression_dict = {unique_tissues[i]: [] for i in range(len(unique_tissues))}\n",
    "    for _, measurement_list in expression.iterrows():\n",
    "        # For each row, create a temporary dictionary\n",
    "        temp_expression_dict = {unique_tissues[i]: [] for i in range(len(unique_tissues))}\n",
    "        for index in range(len(measurement_list)):\n",
    "            temp_expression_dict[tissues[index]].append(measurement_list[index])\n",
    "        # Calculate the mean\n",
    "        for key, values in temp_expression_dict.items():\n",
    "            average_value = np.mean(values)\n",
    "            averaged_expression_dict[key].append(average_value)\n",
    "        \n",
    "    # Convert the Python dictionary into DataFrame\n",
    "    averaged_expression_df = pd.DataFrame.from_dict(averaged_expression_dict)\n",
    "    averaged_expression_df.index = expression.index\n",
    "    return averaged_expression_df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The below test case should return\n",
    "\n",
    "```\n",
    "      A    B\n",
    "0   4.5  2.5\n",
    "1  10.0  7.0\n",
    "```\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      A    B\n",
      "0   4.5  2.5\n",
      "1  10.0  7.0\n"
     ]
    }
   ],
   "source": [
    "test_df = pd.DataFrame([[5,4,3,2],[10,10,6,8]])\n",
    "print(average_by_tissue(test_df, ['A','A','B','B']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate expression for each tissue in the flyatlas data\n",
    "tissue_expression = average_by_tissue(log_expression, tissues)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3114, 34)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tissue_expression.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Task (1b)**\n",
    "\n",
    "WGCNA starts by building a pairwise correlation matrix of genes. Using the matrix you just created, produce an *unsigned* correlation matrix where each cell contains the absolute value of the correlation coefficients.\n",
    "\n",
    "You can calculate the Pearson correlation values yourself, or look up a numpy or scipy function to do so."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ~~ GRADED CELL (2 mark) - complete this cell ~~\n",
    "\n",
    "def calculate_unsigned_correlation(expression):\n",
    "    ''' \n",
    "    Produce the unsigned correlation matrix for a table of gene expression values.\n",
    "    Assume that the columns of the expression matrix are samples and the rows are\n",
    "    genes, and return an array of arrays giving the Pearson correlation between each pair of genes,\n",
    "    in the same order as the rows of the expression table.\n",
    "    '''\n",
    "    \n",
    "    expression_array = np.array(expression)\n",
    "    expression_array_len = len(expression_array)\n",
    "    \n",
    "    # Store every Pearson correlation inside unsigned_correlation_array\n",
    "    unsigned_correlation_array = [[] for i in range(expression_array_len)]\n",
    "    for i in range(expression_array_len):\n",
    "        for j in range(expression_array_len):\n",
    "            pearson_correlation = scipy.stats.pearsonr(expression_array[i], expression_array[j])[0]\n",
    "            unsigned_correlation_array[i].append(abs(pearson_correlation))\n",
    "            \n",
    "    return np.array(unsigned_correlation_array)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The below test case should return (if displayed to a precision of two decimal places)\n",
    "\n",
    "```\n",
    "array([[ 1.  ,  0.95,  0.96,  0.44,  0.3 ,  0.15],\n",
    "       [ 0.95,  1.  ,  1.  ,  0.71,  0.59,  0.46],\n",
    "       [ 0.96,  1.  ,  1.  ,  0.67,  0.54,  0.41],\n",
    "       [ 0.44,  0.71,  0.67,  1.  ,  0.99,  0.95],\n",
    "       [ 0.3 ,  0.59,  0.54,  0.99,  1.  ,  0.99],\n",
    "       [ 0.15,  0.46,  0.41,  0.95,  0.99,  1.  ]])\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[1.  , 0.95, 0.96, 0.44, 0.3 , 0.15],\n",
       "       [0.95, 1.  , 1.  , 0.71, 0.59, 0.46],\n",
       "       [0.96, 1.  , 1.  , 0.67, 0.54, 0.41],\n",
       "       [0.44, 0.71, 0.67, 1.  , 0.99, 0.95],\n",
       "       [0.3 , 0.59, 0.54, 0.99, 1.  , 0.99],\n",
       "       [0.15, 0.46, 0.41, 0.95, 0.99, 1.  ]])"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_df = pd.DataFrame([[ 3.8,  2.7,  4.5],\n",
    "                       [ 4.3,  3.4,  6.2],\n",
    "                       [ 5.3,  4.3,  7. ],\n",
    "                       [ 4.6,  6. ,  7.7],\n",
    "                       [ 5.2,  7.3,  8.8],\n",
    "                       [ 6.2,  8.5,  9.4]], \n",
    "                         columns=['Tissue1', 'Tissue2', 'Tissue3'],\n",
    "                         index=['GeneA', 'GeneB', 'GeneC', 'GeneD', 'GeneE', 'GeneF'])\n",
    "calculate_unsigned_correlation(test_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The below test case should return (if displayed to a precision of two decimal places)\n",
    "\n",
    "```\n",
    "array([[ 1.  ,  0.95,  0.3 ,  0.15],\n",
    "       [ 0.95,  1.  ,  0.59,  0.46],\n",
    "       [ 0.3 ,  0.59,  1.  ,  0.99],\n",
    "       [ 0.15,  0.46,  0.99,  1.  ]])\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[1.  , 0.95, 0.3 , 0.15],\n",
       "       [0.95, 1.  , 0.59, 0.46],\n",
       "       [0.3 , 0.59, 1.  , 0.99],\n",
       "       [0.15, 0.46, 0.99, 1.  ]])"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_df = pd.DataFrame([[ 3.8,  2.7,  4.5],\n",
    "                       [ 4.3,  3.4,  6.2],\n",
    "                       [ 5.2,  7.3,  8.8],\n",
    "                       [ 6.2,  8.5,  9.4]], \n",
    "                         columns=['Tissue1', 'Tissue2', 'Tissue3'],\n",
    "                         index=['GeneA', 'GeneB', 'GeneC', 'GeneD'])\n",
    "calculate_unsigned_correlation(test_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYkAAAD4CAYAAAAZ1BptAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjMsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+AADFEAAATcklEQVR4nO3df4xd5X3n8fenENJ08wMnNhFrux22daTS/EHIiHgVaZeWljhUiqlEWiO1uBFaVyxU7W5UxW3/IEpaKdltGhUpJXWKhYnaEErbxQpQ1yVE2V0F6qFh+VnElFCYYoETE0qFmizZ7/5xn0E343tmrufHnbkz75d0dc/9nuec8zyMM9/5Ps+5J6kqJEka5AdWuwOSpLXLJCFJ6mSSkCR1MklIkjqZJCRJnc5c7Q4st82bN9fExMRqd0OSxsoDDzzwzaraMje+7pLExMQEU1NTq90NSRorSf5xUNzpJklSJ5OEJKmTSUKS1MkkIUnqZJKQJHUySUiSOpkkJEmdTBKSpE4mCUlSp3X3jeulmNh/54Jtnv7Ez46gJ5K0NpgkTlN/IjFhSFrvTBJLYMKQtN6ZJJaJCUPSeuTCtSSpk5XECrCqkLReWElIkjpZSaywubfVWllIGidWEpKkTlYSI+Z6haRxsmAlkWR7knuTPJ7k0SS/1uIfTfJPSR5sr8v6jvnNJNNJnkjyvr74rhabTrK/L35ekvuTPJnki0nOavHXt8/Tbf/Ecg5ekjS/YSqJV4EPV9XfJXkT8ECSo23fp6vq9/obJzkf2AP8BPBvgb9J8o62+zPAzwAzwLEkh6vqMeCT7Vy3JvkscDVwY3t/sap+LMme1u4XljLgtcSqQtJat2AlUVXHq+rv2vbLwOPA1nkO2Q3cWlXfqapvANPARe01XVVPVdV3gVuB3UkC/BRwezv+EHB537kOte3bgUtae0nSCJzWwnWb7nkXcH8LXZfkoSQHk2xqsa3As32HzbRYV/xtwLer6tU58e87V9v/UmsvSRqBoZNEkjcCfw78elX9M73poB8FLgCOA5+abTrg8FpEfL5zze3bviRTSaZOnDgx7zjWqon9d772kqS1Yqi7m5K8jl6C+JOq+guAqnq+b//ngC+1jzPA9r7DtwHPte1B8W8CZyc5s1UL/e1nzzWT5EzgLcDJuf2rqgPAAYDJyclTksi4ca1C0loxzN1NAW4CHq+q3++Ln9vX7OeAR9r2YWBPuzPpPGAH8LfAMWBHu5PpLHqL24erqoB7gSva8XuBO/rOtbdtXwF8ubWXJI3AMJXEe4FfAh5O8mCL/RZwZZIL6E3/PA38CkBVPZrkNuAxendGXVtV3wNIch1wBDgDOFhVj7bzfQS4NcnvAF+nl5Ro759PMk2vgtizhLFKkk5T1tsf5pOTkzU1NbWoY9f6eoBTT5JWSpIHqmpybtzHckiSOpkkJEmdfHbTGPGuJ0mjZiUhSepkkpAkdXK6aUw59SRpFKwkJEmdTBKSpE4mCUlSJ9ck1gHXJyStFCsJSVInk4QkqZPTTeuMU0+SlpOVhCSpk0lCktTJJCFJ6uSaxDrm+oSkpbKSkCR1MklIkjqZJCRJnVyT2CBcn5C0GFYSkqROJglJUienmzYgp54kDctKQpLUySQhSepkkpAkdTJJSJI6uXC9wbmILWk+VhKSpE4LJokk25Pcm+TxJI8m+bUWf2uSo0mebO+bWjxJbkgyneShJBf2nWtva/9kkr198Xcnebgdc0OSzHcNSdJoDFNJvAp8uKp+HNgJXJvkfGA/cE9V7QDuaZ8B3g/saK99wI3Q+4UPXA+8B7gIuL7vl/6Nre3scbtavOsakqQRWHBNoqqOA8fb9stJHge2AruBi1uzQ8BXgI+0+C1VVcB9Sc5Ocm5re7SqTgIkOQrsSvIV4M1V9bUWvwW4HLh7nmtoBbg+IWmu01qTSDIBvAu4H3h7SyCzieSc1mwr8GzfYTMtNl98ZkCcea4xt1/7kkwlmTpx4sTpDEmSNI+hk0SSNwJ/Dvx6Vf3zfE0HxGoR8aFV1YGqmqyqyS1btpzOoZKkeQyVJJK8jl6C+JOq+osWfr5NI9HeX2jxGWB73+HbgOcWiG8bEJ/vGpKkERjm7qYANwGPV9Xv9+06DMzeobQXuKMvflW7y2kn8FKbKjoCXJpkU1uwvhQ40va9nGRnu9ZVc8416BpaYRP773ztJWnjGubLdO8Ffgl4OMmDLfZbwCeA25JcDTwDfLDtuwu4DJgGXgE+BFBVJ5N8HDjW2n1sdhEbuAa4GXgDvQXru1u86xqSpBEY5u6m/8XgdQOASwa0L+DajnMdBA4OiE8B7xwQ/9aga0iSRsPHcmhB3horbVw+lkOS1MkkIUnqZJKQJHUySUiSOrlwrdPiIra0sVhJSJI6mSQkSZ1MEpKkTiYJSVInF661aC5iS+uflYQkqZNJQpLUyekmLQunnqT1yUpCktTJJCFJ6mSSkCR1MklIkjqZJCRJnby7ScvOO52k9cNKQpLUySQhSepkkpAkdXJNQivK9QlpvFlJSJI6mSQkSZ2cbtLIOPUkjR8rCUlSJ5OEJKmTSUKS1GnBJJHkYJIXkjzSF/tokn9K8mB7Xda37zeTTCd5Isn7+uK7Wmw6yf6++HlJ7k/yZJIvJjmrxV/fPk+3/RPLNWitvon9d37fS9LaNEwlcTOwa0D801V1QXvdBZDkfGAP8BPtmD9MckaSM4DPAO8HzgeubG0BPtnOtQN4Ebi6xa8GXqyqHwM+3dpJkkZowSRRVV8FTg55vt3ArVX1nar6BjANXNRe01X1VFV9F7gV2J0kwE8Bt7fjDwGX953rUNu+HbiktZckjchS1iSuS/JQm47a1GJbgWf72sy0WFf8bcC3q+rVOfHvO1fb/1Jrf4ok+5JMJZk6ceLEEoYkSeq32CRxI/CjwAXAceBTLT7oL/1aRHy+c50arDpQVZNVNblly5b5+i1JOg2L+jJdVT0/u53kc8CX2scZYHtf023Ac217UPybwNlJzmzVQn/72XPNJDkTeAvDT3tpzPhFO2ltWlQlkeTcvo8/B8ze+XQY2NPuTDoP2AH8LXAM2NHuZDqL3uL24aoq4F7ginb8XuCOvnPtbdtXAF9u7SVJI7JgJZHkC8DFwOYkM8D1wMVJLqA3/fM08CsAVfVoktuAx4BXgWur6nvtPNcBR4AzgINV9Wi7xEeAW5P8DvB14KYWvwn4fJJpehXEniWPVpJ0WhZMElV15YDwTQNis+1/F/jdAfG7gLsGxJ+id/fT3Pi/Ah9cqH+SpJXjN64lSZ18CqzWHBexpbXDSkKS1MkkIUnqZJKQJHVyTUJrmusT0uqykpAkdTJJSJI6mSQkSZ1ck9DYcH1CGj0rCUlSJ5OEJKmTSUKS1MkkIUnq5MK1xpKL2NJoWElIkjpZSWjsWVVIK8dKQpLUySQhSepkkpAkdTJJSJI6uXCtdcVFbGl5WUlIkjqZJCRJnUwSkqROrklo3XJ9Qlo6KwlJUieThCSpk9NN2hCcepIWZ8FKIsnBJC8keaQv9tYkR5M82d43tXiS3JBkOslDSS7sO2Zva/9kkr198Xcnebgdc0OSzHcNSdLoDDPddDOwa05sP3BPVe0A7mmfAd4P7GivfcCN0PuFD1wPvAe4CLi+75f+ja3t7HG7FriGJGlEFkwSVfVV4OSc8G7gUNs+BFzeF7+leu4Dzk5yLvA+4GhVnayqF4GjwK62781V9bWqKuCWOecadA1J0ogsdk3i7VV1HKCqjic5p8W3As/2tZtpsfniMwPi811DWhLXJ6ThLffdTRkQq0XET++iyb4kU0mmTpw4cbqHS5I6LDZJPN+mimjvL7T4DLC9r9024LkF4tsGxOe7ximq6kBVTVbV5JYtWxY5JEnSXItNEoeB2TuU9gJ39MWvanc57QRealNGR4BLk2xqC9aXAkfavpeT7Gx3NV0151yDriFJGpEF1ySSfAG4GNicZIbeXUqfAG5LcjXwDPDB1vwu4DJgGngF+BBAVZ1M8nHgWGv3saqaXQy/ht4dVG8A7m4v5rmGtGxcn5Dmt2CSqKorO3ZdMqBtAdd2nOcgcHBAfAp454D4twZdQ5I0Oj6WQ5LUycdySI1TT9KprCQkSZ1MEpKkTiYJSVIn1ySkAVyfkHqsJCRJnUwSkqROTjdJC3DqSRuZlYQkqZNJQpLUyekm6TQ49aSNxkpCktTJJCFJ6uR0k7RITj1pI7CSkCR1MklIkjqZJCRJnVyTkJaB6xNar6wkJEmdTBKSpE5ON0nLzKknrSdWEpKkTlYS0gqyqtC4s5KQJHUySUiSOjndJI2IU08aR1YSkqROJglJUienm6RV4NSTxsWSKokkTyd5OMmDSaZa7K1JjiZ5sr1vavEkuSHJdJKHklzYd569rf2TSfb2xd/dzj/djs1S+itJOj3LMd30k1V1QVVNts/7gXuqagdwT/sM8H5gR3vtA26EXlIBrgfeA1wEXD+bWFqbfX3H7VqG/kprysT+O197SWvNSqxJ7AYOte1DwOV98Vuq5z7g7CTnAu8DjlbVyap6ETgK7Gr73lxVX6uqAm7pO5ckaQSWmiQK+OskDyTZ12Jvr6rjAO39nBbfCjzbd+xMi80XnxkQP0WSfUmmkkydOHFiiUOSJM1a6sL1e6vquSTnAEeT/P08bQetJ9Qi4qcGqw4ABwAmJycHtpHGwdwpJxe1tdqWVElU1XPt/QXgL+mtKTzfpopo7y+05jPA9r7DtwHPLRDfNiAuSRqRRSeJJP8myZtmt4FLgUeAw8DsHUp7gTva9mHgqnaX007gpTYddQS4NMmmtmB9KXCk7Xs5yc52V9NVfeeSNgQXtbXaljLd9HbgL9tdqWcCf1pVf5XkGHBbkquBZ4APtvZ3AZcB08ArwIcAqupkko8Dx1q7j1XVybZ9DXAz8Abg7vaSJI1IejcOrR+Tk5M1NTW1qGP9a03jwrUKLbckD/R9leE1PpZDktTJx3JIY8jHemhUrCQkSZ2sJKQxZ1WhlWQlIUnqZCUhrSNWFVpuVhKSpE5WEtI6ZVWh5WAlIUnqZCUhbQBWFVosKwlJUicrCWmDsarQ6TBJSBuYCUMLcbpJktTJSkISYFWhwUwSkk5hwtAsk4SkeZkwNjbXJCRJnawkJA3NqmLjMUlIWhQTxsZgkpC0ZCaM9cskIWlZmTDWF5OEpBVjwhh/JglJI2HCGE8mCUkjZ8IYHyYJSauqP2H0M3msDSYJSWuS1cbaYJKQtOZ1VRtgAllpJglJY83pqpVlkpC0Lpk8lseaTxJJdgF/AJwB/HFVfWKVuyRpjM03dTVrXBLJ3LGsRL/XdJJIcgbwGeBngBngWJLDVfXY6vZM0no2TCLpNy5JZTHWdJIALgKmq+opgCS3ArsBk4SkNeN0k8pSjDohrfUksRV4tu/zDPCeuY2S7AP2tY//kuSJRV5vM/DNRR47rhzzxuCY14l8ct59SxnzjwwKrvUkkQGxOiVQdQA4sOSLJVNVNbnU84wTx7wxOOaNYSXGvNb/n+lmgO19n7cBz61SXyRpw1nrSeIYsCPJeUnOAvYAh1e5T5K0Yazp6aaqejXJdcARerfAHqyqR1fwkkueshpDjnljcMwbw7KPOVWnTPFLkgSs/ekmSdIqMklIkjptyCSRZFeSJ5JMJ9k/YP/rk3yx7b8/ycToe7m8hhjzf03yWJKHktyTZOA90+NkoTH3tbsiSSUZ69slhxlvkp9vP+dHk/zpqPu43Ib4d/3DSe5N8vX2b/uy1ejnckpyMMkLSR7p2J8kN7T/Jg8luXBJF6yqDfWitwD+D8C/A84C/g9w/pw2/xn4bNveA3xxtfs9gjH/JPBDbfuajTDm1u5NwFeB+4DJ1e73Cv+MdwBfBza1z+esdr9HMOYDwDVt+3zg6dXu9zKM+z8AFwKPdOy/DLib3vfMdgL3L+V6G7GSeO1RH1X1XWD2UR/9dgOH2vbtwCVJBn2xb1wsOOaqureqXmkf76P3nZRxNszPGeDjwH8D/nWUnVsBw4z3PwGfqaoXAarqhRH3cbkNM+YC3ty238I6+J5VVX0VODlPk93ALdVzH3B2knMXe72NmCQGPepja1ebqnoVeAl420h6tzKGGXO/q+n9JTLOFhxzkncB26vqS6Ps2AoZ5mf8DuAdSf53kvvaE5bH2TBj/ijwi0lmgLuAXx1N11bV6f7vfV5r+nsSK2SYR30M9TiQMTL0eJL8IjAJ/McV7dHKm3fMSX4A+DTwy6Pq0Aob5md8Jr0pp4vpVYr/M8k7q+rbK9y3lTLMmK8Ebq6qTyX598Dn25j/38p3b9Us6++vjVhJDPOoj9faJDmTXpk6X3m31g31eJMkPw38NvCBqvrOiPq2UhYa85uAdwJfSfI0vbnbw2O8eD3sv+s7qur/VtU3gCfoJY1xNcyYrwZuA6iqrwE/SO/Bf+vZsj7OaCMmiWEe9XEY2Nu2rwC+XG1FaEwtOOY29fJH9BLEuM9VwwJjrqqXqmpzVU1U1QS9dZgPVNXU6nR3yYb5d/0/6N2gQJLN9KafnhppL5fXMGN+BrgEIMmP00sSJ0bay9E7DFzV7nLaCbxUVccXe7INN91UHY/6SPIxYKqqDgM30StLp+lVEHtWr8dLN+SY/zvwRuDP2hr9M1X1gVXr9BINOeZ1Y8jxHgEuTfIY8D3gN6rqW6vX66UZcswfBj6X5L/Qm3L55TH/g48kX6A3Zbi5rbVcD7wOoKo+S2/t5TJgGngF+NCSrjfm/70kSStoI043SZKGZJKQJHUySUiSOpkkJEmdTBKSpE4mCUlSJ5OEJKnT/wcrN3S2HfFx0wAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Calculate the correlation matrix for the flyatlas data\n",
    "unsigned_correlation = calculate_unsigned_correlation(tissue_expression)\n",
    "_ = plt.hist(unsigned_correlation.flatten(), bins=100)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "**TASK (1c)** Why are we using an unsigned correlation matrix instead of a signed correlation matrix? (max 50 words)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# ~~ GRADED CELL (1 mark) - your answer here --\n",
    "In this WGCNA, unsigned correlation matrix is used because it is assumed that strong negative correlation indicates connection between two genes as much as strong positive correlation. Since only the magnitude matters in this case, positive/negative sign is not required."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "### Task 2 - Building an adjacency matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To use the correlation matrix to create a network, we will transform it into an adjacency matrix. You will create two types of adjacency matrix, a binary adjacency matrix and a weighted adjacency matrix."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Task (2a)**\n",
    "\n",
    "To create the binary adjacency matrix, transform the correlation matrix such that every correlation greater than or equal to a given threshold value is considered adjacent (represented by a 1 in the matrix), and every correlation below that value is considered not adjacent (represented by a 0). Set the diagonal of the adjacency matrix to 0, so that we don't consider a node to be adjacent to itself."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ~~ GRADED CELL (1 marks) - complete this cell ~~\n",
    "\n",
    "def calculate_binary_adjacencies(correlation, threshold):\n",
    "    '''\n",
    "    Given a correlation matrix between genes of shape (N,N),\n",
    "    return the corresponding binary adjacency matrix of shape (N,N),\n",
    "    where correlation values are above the given threshold.\n",
    "    '''\n",
    "    N = len(correlation)\n",
    "    binary_adjacency_matrix = [[] for i in range(N)]\n",
    "    for i in range(N):\n",
    "        for j in range(N):\n",
    "            if i==j:\n",
    "                # Set the diagonal to 0\n",
    "                binary_adjacency_matrix[i].append(0)\n",
    "            else:\n",
    "                if correlation[i][j] >= threshold:\n",
    "                    # Adjacent if passing the threshold\n",
    "                    binary_adjacency_matrix[i].append(1)\n",
    "                else:\n",
    "                    binary_adjacency_matrix[i].append(0)\n",
    "    \n",
    "    return np.array(binary_adjacency_matrix)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The below test case should return (if displayed to a precision of two decimal places)\n",
    "\n",
    "```\n",
    "array([[ 0.,  1.,  0.,  0.],\n",
    "       [ 1.,  0.,  1.,  0.],\n",
    "       [ 0.,  1.,  0.,  1.],\n",
    "       [ 0.,  0.,  1.,  0.]])\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0, 1, 0, 0],\n",
       "       [1, 0, 1, 0],\n",
       "       [0, 1, 0, 1],\n",
       "       [0, 0, 1, 0]])"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_corr = np.array([[ 1.  ,  0.95,  0.3 ,  0.15],\n",
    "       [ 0.95,  1.  ,  0.59,  0.46],\n",
    "       [ 0.3 ,  0.59,  1.  ,  0.99],\n",
    "       [ 0.15,  0.46,  0.99,  1.  ]])\n",
    "calculate_binary_adjacencies(test_corr, 0.5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The below test case should return (if displayed to a precision of two decimal places)\n",
    "\n",
    "```\n",
    "array([[ 0.,  1.,  0.,  0.],\n",
    "       [ 1.,  0.,  0.,  0.],\n",
    "       [ 0.,  0.,  0.,  1.],\n",
    "       [ 0.,  0.,  1.,  0.]])\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0, 1, 0, 0],\n",
       "       [1, 0, 0, 0],\n",
       "       [0, 0, 0, 1],\n",
       "       [0, 0, 1, 0]])"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_corr = np.array([[ 1.  ,  0.95,  0.3 ,  0.15],\n",
    "       [ 0.95,  1.  ,  0.59,  0.46],\n",
    "       [ 0.3 ,  0.59,  1.  ,  0.99],\n",
    "       [ 0.15,  0.46,  0.99,  1.  ]])\n",
    "calculate_binary_adjacencies(test_corr, 0.6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate the binary adjacency matrix for the flyatlas data\n",
    "adjacency_binary = calculate_binary_adjacencies(unsigned_correlation, 0.85)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Task (2b)**\n",
    "\n",
    "Calculate the connectivity of the adjacency matrix by dividing the total number of edges by the number of possible edges."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ~~ GRADED CELL (1 mark) - complete this cell ~~\n",
    "\n",
    "def calculate_connectivity(adjacency):\n",
    "    '''\n",
    "    Calculate the number of edges that exist in a given binary adjacency matrix,\n",
    "    divided by the total number of possible edges between all nodes.\n",
    "    '''\n",
    "    N = len(adjacency)\n",
    "    number_of_edges = 0\n",
    "    # Iterate through adjacency matrix to calculate number_of_edges\n",
    "    for i in range(N):\n",
    "        for j in range(N):\n",
    "            if adjacency[i][j] == 1:\n",
    "                number_of_edges += 1\n",
    "    # Number of possible edges is the number of every cell inside adjacency matrix except the diagonal\n",
    "    number_of_possible_edges = (N*N) - N\n",
    "    return number_of_edges / number_of_possible_edges\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.5"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Should return 0.5\n",
    "calculate_connectivity(np.array([[ 0.,  1.,  0.,  0.],\n",
    "                                   [ 1.,  0.,  1.,  0.],\n",
    "                                   [ 0.,  1.,  0.,  1.],\n",
    "                                   [ 0.,  0.,  1.,  0.]]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.3333333333333333"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Should return 0.33\n",
    "calculate_connectivity(np.array([[ 0.,  1.,  0.,  0.],\n",
    "                                   [ 1.,  0.,  0.,  0.],\n",
    "                                   [ 0.,  0.,  0.,  1.],\n",
    "                                   [ 0.,  0.,  1.,  0.]]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0020161169694452646"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "calculate_connectivity(adjacency_binary)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Task 2(c)**\n",
    "\n",
    "The weighted adjacency matrix can be created by raising the correlation matrix to some power. Write a function that raises the correlation matrix to some power, `beta`, and sets the diagonal to `0`. For the rest of the assignment we will use `beta = 4` but your function should accept any integer."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ~~ GRADED CELL (2 marks) - complete this cell ~~\n",
    "\n",
    "def calculate_weighted_adjacencies(correlation, beta):\n",
    "    '''\n",
    "    Given a correlation matrix between genes of shape (N,N),\n",
    "    return the corresponding binary adjacency matrix of shape (N,N),\n",
    "    where we use a power-law soft threshold with parameter beta.\n",
    "    '''\n",
    "    N = len(correlation)\n",
    "    weighted_adjacency_matrix = [[] for i in range(N)]\n",
    "    for i in range(N):\n",
    "        for j in range(N):\n",
    "            if i==j:\n",
    "                # Set the diagonals to 0\n",
    "                weighted_adjacency_matrix[i].append(0)\n",
    "            else:\n",
    "                # Raise the correlation value to the power of beta\n",
    "                weighted_adjacency_matrix[i].append(correlation[i][j] ** beta)\n",
    "    \n",
    "    return weighted_adjacency_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The below test case should return (if displayed to a precision of two decimal places)\n",
    "\n",
    "```\n",
    "array([[ 0.  ,  0.9 ,  0.09,  0.02],\n",
    "       [ 0.9 ,  0.  ,  0.35,  0.21],\n",
    "       [ 0.09,  0.35,  0.  ,  0.98],\n",
    "       [ 0.02,  0.21,  0.98,  0.  ]])\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[0, 0.9025, 0.09, 0.0225],\n",
       " [0.9025, 0, 0.34809999999999997, 0.2116],\n",
       " [0.09, 0.34809999999999997, 0, 0.9801],\n",
       " [0.0225, 0.2116, 0.9801, 0]]"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_corr = np.array([[ 1.  ,  0.95,  0.3 ,  0.15],\n",
    "       [ 0.95,  1.  ,  0.59,  0.46],\n",
    "       [ 0.3 ,  0.59,  1.  ,  0.99],\n",
    "       [ 0.15,  0.46,  0.99,  1.  ]])\n",
    "calculate_weighted_adjacencies(test_corr, 2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The below test case should return (if displayed to a precision of two decimal places)\n",
    "\n",
    "```\n",
    "array([[ 0.  ,  0.86,  0.03,  0.  ],\n",
    "       [ 0.86,  0.  ,  0.21,  0.1 ],\n",
    "       [ 0.03,  0.21,  0.  ,  0.97],\n",
    "       [ 0.  ,  0.1 ,  0.97,  0.  ]])\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[0, 0.8573749999999999, 0.026999999999999996, 0.0033749999999999995],\n",
       " [0.8573749999999999, 0, 0.20537899999999998, 0.097336],\n",
       " [0.026999999999999996, 0.20537899999999998, 0, 0.970299],\n",
       " [0.0033749999999999995, 0.097336, 0.970299, 0]]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_corr = np.array([[ 1.  ,  0.95,  0.3 ,  0.15],\n",
    "       [ 0.95,  1.  ,  0.59,  0.46],\n",
    "       [ 0.3 ,  0.59,  1.  ,  0.99],\n",
    "       [ 0.15,  0.46,  0.99,  1.  ]])\n",
    "calculate_weighted_adjacencies(test_corr, 3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate the weighted adjacency matrix for the flyatlas data\n",
    "adjacency_weighted = calculate_weighted_adjacencies(unsigned_correlation, 4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Task 2(d)**\n",
    "\n",
    "How do you expect the network connectivity would change if the threshold for the binary adjacency matrix is increased or decreased? (max 50 words)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# ~~ GRADED CELL (1 mark) - your answer here --\n",
    "If the threshold in binary adjacency matrix is increased, then less gene pairs will be considered as connected. Hence, the number of edges is decreased and the network connectivity value will decrease. Vice versa. Hence, threshold has a negative correlation with network connectivity"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Task 3 - Dimensionality Reduction"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this task we will be performing Priciple Components Analysis to determine which gene in the first principle component has the highest contribution to the variance."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Task 4(a)**\n",
    "\n",
    "Perform a Principle Componanets Analysis on the log_expression matrix with the correct number of components and print the explained variance by component list."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Explained Variance:  [0.17 0.13 0.08 0.07 0.07 0.06 0.05 0.03 0.03 0.03 0.02 0.02 0.02 0.02\n",
      " 0.01 0.01 0.01 0.01 0.01 0.01 0.01 0.01 0.01 0.01 0.01]\n",
      "Number of components to reach ~90% explained variance:  25\n"
     ]
    }
   ],
   "source": [
    "# ~~ GRADED CELL (1 mark) - complete this cell ~~\n",
    "pca = PCA(n_components=25)\n",
    "expression_pca = pca.fit_transform(log_expression.values.T)\n",
    "print(\"Explained Variance: \", pca.explained_variance_ratio_)\n",
    "\n",
    "# Determine the appropriate number of components for the PCA\n",
    "i = 0\n",
    "cumulative_explained_variance = 0\n",
    "while cumulative_explained_variance < 0.9:\n",
    "    cumulative_explained_variance += pca.explained_variance_ratio_[i]\n",
    "    i += 1\n",
    "print(\"Number of components to reach ~90% explained variance: \", i)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Task 4(b)**\n",
    "\n",
    "Print the gene that contributes most to the first eigen vector (the first principle component) of the PCA. The word limit on the second cell is 50 words."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene that contributes the most to 1st eigenvector:  Adult Brain\n"
     ]
    }
   ],
   "source": [
    "# ~~ GRADED CELL (2 marks) - complete this cell ~~\n",
    "maximum_index = 0\n",
    "maximum_value = expression_pca[0][0]\n",
    "# Use expression_pca to find the maximum value inside column 0 (1st principle component)\n",
    "for row in range(len(expression_pca)):\n",
    "    if expression_pca[row][0] > maximum_value:\n",
    "        maximum_index = row\n",
    "        maximum_value = expression_pca[row][0]\n",
    "print(\"Gene that contributes the most to 1st eigenvector: \", tissues[maximum_index])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Task 4 - Graph Metrics"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Graph metrics are important parameters to assist in characterising a network as a whole or even the relative importance of specific nodes in a network and could give us hints regarding their importance."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Task 5(a)**\n",
    "\n",
    "Normalised Degree Centrality\n",
    "\n",
    "Describe an algorithm in pseudocode that returns the normalised degree centrality of a node (degree divided by the maximum node degree in the graph), receiving as parameters a node index **i** and its binary adjacency matrix **m**."
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# ~~ GRADED CELL (2 marks) - your answer here --\n",
    "function normalisedDegreeCentrality(i, m)\n",
    "    Set degree to 0\n",
    "    Set maximumDegree to 0\n",
    "    Set numberOfRow = number of row in m\n",
    "    Set numberOfColumn = number of column in m\n",
    "    For rowIndex = 0 to numberOfRow - 1\n",
    "        Set currentDegree to 0\n",
    "        For columnIndex = 0 to numberOfColumn - 1\n",
    "            Set currentDegree to currentDegree + m[rowIndex][columnIndex]\n",
    "        if rowIndex equals to i Then\n",
    "            Set degree to currentDegree\n",
    "        if currentDegree is bigger than maximumDegree Then\n",
    "            Set maximumDegree to currentDegree\n",
    "    Set degreeCentrality to degree divided by maximumDegree\n",
    "    return degreeCentrality"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Task 5(b)**\n",
    "\n",
    "Closeness Centrality\n",
    "\n",
    "Describe an algorithm in pseudocode that returns the closeness centrality of a node, receiving as parameters a node index **i** and its binary adjacency matrix **m**. As part of your answer you can assume the function *min_dist(a,b,m)* is available. This function will return the minimum distance between nodes a and b in a graph represented by the adjecendy matrix m."
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# ~~ GRADED CELL (2 marks) - your answer here --\n",
    "function closenessCentrality(i, m)\n",
    "    Set numberOfNodes to number of row in m\n",
    "    Set sumDistance to 0\n",
    "    for node = 0 to numberOfNodes - 1\n",
    "        if node not equals to i Then\n",
    "            Set sumDistance to sumDistance + min_dist(i, node, m)\n",
    "    Set centrality to 1 divided by sumDistance\n",
    "    return centrality"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Task 5(c)**\n",
    "\n",
    "Clustering Coefficient and Average Path Length are two important properties to distinguish between different network types.\n",
    "\n",
    "Consider that you are working with a particular biological network. Describe briefly, with your own words, how would you use these properties to verify whether your network is consistent with a Random Network, Small-World Network or Regular Lattice Network. (Maximum 150 words).\n"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# ~~ GRADED CELL (4 marks) - your answer here --\n",
    "The main difference between these networks are essentially its probability to randomly assigned/reassigned edges (p). Regular Lattice Network has p=0, Random Network has p=1, and Small-World Network in between 0 and 1. As the value of p increases, the clustering coefficient (C) will slowly decrease because the probability that two neighbour nodes are connected is lower. Also, the average path length (L) will rapidly decrease as the value of p increases, mainly because there is a higher probability to find shorter paths between two nodes.\n",
    "Hence, if a biological network has a high C and high L values, then it is consistent with a Regular Lattice Network. If the network has a low C and low L values, then it is consistent with a Random Network. However, if the network has a relatively high C but low L value, then it is a Small-World Network."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Part II\n",
    "<br>\n",
    "Consider the following scenario. As a Bioinformatician you were given the task of designing a computational approach to organise (in a efficient way) genomic data of over 1,000,000 SARS-CoV2 samples extracted from patients worldwide, as well as patient information from hospitalisations to enable quick data access and to enable a better understanding of variants associated with poor patient outcomes.\n",
    "\n",
    "Each data point is composed of:\n",
    "* A patient ID and further patient information (a list of procedures performed).\n",
    "* A list of variants (variant IDs) identified by sequencing the sample collected from the patient.\n",
    "\n",
    "\n",
    "You are interested in building a single index, using both **patient ID** and **variant ID** to retrieve, include or remove information (list of procedures performed) in O(1) time. Your approach also needs to be memory efficient, O(n), and also built in O(n) time.\n",
    "\n",
    "However, given the ever growing number of available sequences/patients and observed variants using (**patient ID + variant ID**) as a single key becomes impractical for traditional hash tables (considering all possible combinations for both).\n",
    "\n",
    "**a)** Describe with **your own words** how you would implement a computational approach to achieve these requirements using or adapting an abstract data type described during our lectures (you are allowed to draw your data structure if that helps). (Maximum 150 words) "
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# ~~ GRADED CELL (10 marks) - your answer here --\n",
    "In this scenario, we are going to use string aggregate of patientID and variantID as the search key. The index file is a list of procedures performed. To achieve this, I propose using a hash table with Burrows-Wheeler transform (BWT) as a hash function. Using BWT index as the search key, it will mitigate the growing number of possible patientID and variantID combinations. Using this approach, it satisfies the required space complexity (O(n)) and time complexity during precomputation (O(n))."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**(b)** Additionally, also describe in **pseudocode** the functions *search(), insert() and remove()* for your abstract data type. <br>**Python code is not acceptable**. (Maximum of 150 words)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# ~~ GRADED CELL (10 marks) - your answer here --\n",
    "# Assuming key = patientID + variantID and the hash table is stored inside a global array called hashTable\n",
    "\n",
    "function bwtTransform(key)\n",
    "    Set key to key + \"$\"\n",
    "    Set keyLength to length of key\n",
    "    \n",
    "    # 1: Generate bwtMatrix\n",
    "    Initialize bwtMatrix[keyLength]\n",
    "    Set bwtMatrix[0] to key\n",
    "    for i=1 to keyLength - 1\n",
    "        Set keyRotation to bwtMatrix[i-1] + bwtMatrix[i-1][0]\n",
    "        Remove bwtMatrix[0]\n",
    "        Set bwtMatrix[i] to keyRotation\n",
    "    \n",
    "    # 2: Sort the bwtMatrix based on alphabet (ascending order)\n",
    "    sort(bwtMatrix)\n",
    "    \n",
    "    # 3: Retrieve the last column from each row to bwtKey\n",
    "    for i=0 to keyLength-1\n",
    "        Set bwtKey to bwtKey + bwtMatrix[i][keyLength-1]\n",
    "    \n",
    "    return bwtKey\n",
    "\n",
    "function search(key)\n",
    "    Set hashIndex to bwtTransform(key)\n",
    "    return hashTable[hashIndex]\n",
    "\n",
    "function insert(key, value)\n",
    "    Set hashIndex to bwtTransform(key)\n",
    "    while hashTable[hashIndex] is equals to NULL then\n",
    "        Set hashTable[hashIndex] equals to (key, value)\n",
    "\n",
    "function remove(key)\n",
    "    Set hashIndex to bwtTransform(key)\n",
    "    Set hashTable[hashIndex] to NULL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
